\documentclass[10]{beamer}
\include{pres_preamble}

\title{Comparison of long and short amplicon metabarcoding for determination of fungal species diversity and spatial turnover}
\subtitle{in an ectomycorrhizal West African woodland}
\author{\mbox{\textbf{Brendan Furneaux}\inst{1}} \and \mbox{Mohammad Bahram\inst{2}} \and \mbox{Nourou Yorou\inst{3}} \and \mbox{Anna Rosling\inst{1}} \and \mbox{Martin Ryberg\inst{1}}}
\institute{\inst{1}Uppsala University \and \inst{2}Swedish University of Agricultural Sciences \and \inst{3}University of Parakou}
\date{13 August, 2019}
\logo{\includegraphics[width=2.5cm]{Uppsala_University_logo}}
\setbeamersize{text margin left=5mm,text margin right=5mm} 

\begin{document}

\frame{\titlepage}

\logo{}

<<setup, include=FALSE>>=
knitr::opts_chunk$set(echo = FALSE, device = "tikz", cache = TRUE)
library(drake)
library(magrittr)
library(tidyverse)
library(ggplot2)
library(scales)
library(ggtree)
library(castor)
theme_set(theme_bw(base_size = 24) + theme(legend.direction = "vertical"))
loadd(
  bc_corr_long,
  bc_corr_long2,
  bc_corr_short,
  bc_corr_short2,
  # unif_corr,
  # unif_corr2,
  qstats_join,
  datasets,
  # longtree#,
  # taxon_LSUcons_rdp
  )
reads <- qstats_join %>%
  mutate_at("region", factor,
            levels = c("ITS2", "ITS1", "short", "ITS", "LSU", "long")) %>%
  mutate_at("dataset", factor,
            levels = c("short-ion", "short-pacbio", "long-pacbio")) %>%
  group_by(dataset, plate, region)
reads_subset <- reads %>%
  do(if (nrow(.) > 10000) sample_n(., 10000) else .)
reverselog_trans <- function(base = exp(1)) {
    trans <- function(x) -log(x, base)
    inv <- function(x) base^(-x)
    trans_new(paste0("reverselog-", format(base)), trans, inv, 
              log_breaks(base = base), 
              domain = c(1e-100, Inf))
}
@


\begin{frame}{Outline}
\Large
\begin{itemize}
  \item Long amplicons for metabarcoding
  \item Study site \& design
  \item (early) Results
\end{itemize}
\end{frame}

\begin{frame}{rDNA Regions}
\vskip -1ex

\includegraphics[width=\linewidth]{rDNA}

\vskip 2ex

\begin{overprint}
\onslide<1>
\small
\begin{tabulary}{\linewidth}{r|L|L|L}
& SSU (16S/18S) & ITS & LSU (23S/26S/28S) \\ \midrule
Variation &
  Highly conserved w/variable regions &
  Variable (5.8S conserved) &
  Conserved w/variable regions \\ \noalign{\smallskip}
Uses & 
  Prokaryota, broad Eukaryota &
  Official barcode for Fungi &
  Phylogenetics \\ \noalign{\smallskip}
Alignable &
  Yes &
  No &
  Yes \\ \noalign{\smallskip}
Databases &
  Silva, RDP (Prokaryotes only) &
  Unite, Warcup &
  Silva, RDP (Fungi) \\ 
\end{tabulary}

\onslide<2>
\begin{columns}[t]
  \begincol{0.3\linewidth}
    \begin{block}{Silva 132 (2017)}
      \small
      \textbf{SSU Ref:}
      \begin{itemize}
        \footnotesize
        \tightlist
        \item 2M seqs
        \item 700k "Ref NR 99" (clusters)
        \item 16k Fungi, 60k other Eukaryota, 625k Prokaryota
      \end{itemize}
      \textbf{LSU Ref:}
      \begin{itemize}
        \footnotesize
        \tightlist
        \item 200k seqs
        \item 5k Fungi, 24k other Eukaryota, 170k Prokaryota
      \end{itemize}
    \end{block}
  \endcol
  \begincol{0.3\linewidth}
    \begin{block}{Unite 8.0 (2019)}
    \small
    \textbf{ITS:}
      \begin{itemize}
        \footnotesize
        \tightlist
        \item 800k seqs
        \item 120k "species hypotheses" (clusters)
        \item 70k Fungi, 40k plants, 10k other Eukaryota
        \item some sequences include LSU
      \end{itemize}
    \end{block}
  \endcol
  \begincol{0.3\linewidth}
    \small
    \begin{block}{RDP 11 (2016)}
      \small
      \textbf{LSU training set:}\\
      \begin{itemize}
        \footnotesize
        \tightlist
        \item 11k seqs
        \item 8k Fungi, 3k other Eukaryota
        \item annotated to genus
      \end{itemize}
      \textbf{SSU dataset:}
      \begin{itemize}
        \footnotesize
        \tightlist
        \item Prokaryotes only
      \end{itemize}
      \textbf{Warcup Fungal ITS training set:}
      \begin{itemize}
        \footnotesize
        \tightlist
        \item 18k seqs, all Fungi
      \end{itemize}
    \end{block}
  \endcol
\end{columns}
\center The three databases don't use the same taxonomy!
\end{overprint}
\end{frame}

%\begin{frame}{Outline}
%\tableofcontents
%\end{frame}

\begin{frame}{Why long amplicons?}

\begin{itemize}
 \item Combine ITS (as barcode) with LSU/SSU (phylogenetic inference)
 \item Long read technologies are more tolerant of length variation
 \begin{itemize}
    \item PCR length bias still an issue!
 \end{itemize}
 \item Very conserved primer sites reduce primer bias.
 \item (Tedersoo et al., 2018)
\end{itemize}
\end{frame}

\begin{frame}{Why not long amplicons?}
\begin{itemize}
  \item Read quality
    \begin{itemize}
      \item PacBio circular consensus gives good quality up to a few kb
      \item Oxford Nanopore is still problematic
    \end{itemize}
  \item Chimera formation in PCR.
    \begin{itemize}
      \item Alternation of conserved and variable regions is a recipe for chimera formation.
    \end{itemize}
  \item Cost -- for the same budget, Illumina gives many more samples and/or greater depth.
  \item (Kennedy et al. 2018)

\end{itemize}
\end{frame}

\begin{frame}{Does it matter?}
\begin{itemize}
  \item Do a simple ecological study using different methods, see if the results change.
  \item Measuring spatial turnover in soil fungal community.
\end{itemize}
\end{frame}

\begin{frame}{Study site}
\vskip 1.5em
\begincols
  \begincol{0.55\linewidth}
    \begin{figure}
      \begin{overprint}
      \onslide<1>
        \includegraphics{soudanian_benin}
      \onslide<2>
        \includegraphics{soudanian_benin_oueme}
      \end{overprint}
      \caption{Location of study sites in Bènin.}
    \end{figure}
  \endcol

  \begincol{0.45\linewidth}
  \begin{itemize}
    \tightlist
    \item West Soudanian woodlands in central Bènin
    \item 10° N Latitude
    \item Two sites 30km apart
    \item Dominated by \emph{Isoberlinia doka} (Caesalpinoideae; ECM host)
    \item Many AM trees and shrubs, plus three more ECM.
    \item Fire \& termites
  \end{itemize}
  \endcol
\endcols
\end{frame}

\begin{frame}{Soudanian Woodlands}
  \begin{overprint}
    \onslide<1>
      \includegraphics{Foret_Oueme_wide}
    \onslide<2>
      \includegraphics{woodland}
  \end{overprint}
\end{frame}

\begin{frame}{Sampling}
  \begin{itemize}
    \item \SI{25}{\m} transects.
    \item Soil samples taken every 1 m.
    \item $\SI{5}{\cm}\times\SI{5}{\cm}\times\SI{5}{\cm}$ sample, homogenize in plastic bag, then subsample \SI{50}{\mg}
    \item Lysis \& preservation in field
    \item Same locations in two years -- 2015 and 2016 (25 points × 2 years × 2 sites = 100 samples) 
  \end{itemize}
\end{frame}

\begin{frame}{Sequencing}
  \framesubtitle{Which amplicon?}
  \includegraphics[width=\linewidth]{rDNA_amplicons}
  \vfill
  \begin{columns}[t]
  \begincol{0.5\linewidth}
    \begin{block}{short amplicon}
      \begin{itemize}
        \tightlist
        \item ITS2
        \item primers gITS7 -- ITS4
        \item 250--350 bp
        \item Sequence same amplicon library with Illumina MiSeq and PacBio RSII
        \item<2-> Don't have MiSeq data yet!
       \end{itemize}
     \end{block}
  \endcol
  \begincol{0.5\linewidth}
     \begin{block}{long amplicon}
        \begin{itemize}
          \tightlist
          \item Full ITS + partial LSU
          \item Primers ITS1 -- LR5
          \item 1200 -- 1800 bp
          \item PacBio RSII
        \end{itemize}
      \end{block}
  \endcol
\end{columns}
\end{frame}
% 
% \begin{frame}{Sequencing}
%   \begin{block}{The plan}
%   Three sequencing runs:
%   \begin{itemize}
%     \item Ion Proton S5, short amplicon
%     \begin{itemize}
%       \item Pool all samples on one chip
%     \end{itemize}
%     \item PacBio RSII, short amplicon
%     \begin{itemize}
%       \item Same amplicon library as IonTorrent.
%       \item Two sequencing libraries, two SMRT cells per library.
%     \end{itemize}
%     \item PacBio RSII, long amplicon
%     \begin{itemize}
%       \item Two sequencing libraries, four SMRT cells per library.
%     \end{itemize}
%   \end{itemize}
%   \end{block}
% \end{frame}
% 
% \begin{frame}{Sequencing}
%   \begin{block}{What happened}
%   Three sequencing runs:
%   \begin{itemize}
%     \item \sout{IonTorrent, short amplicon}
%     \begin{itemize}
%       \item Accidentally used primers with IonTorrent adapters built in.
%       \item Lost some of the multiplexing barcodes.
%       \item Can't demultiplex the samples --- can only look at pool of all reads.
%     \end{itemize}
%     \item PacBio, short amplicon
%     \begin{itemize}
%       \item OK
%     \end{itemize}
%     \item PacBio, long amplicon
%     \begin{itemize}
%       \item OK
%     \end{itemize}
%   \end{itemize}
%   \end{block}
% \end{frame}

\begin{frame}{Bioinformatics}
  \begin{itemize}
    \item Redo RSII circular consensus basecalling using Sequel algorithm (Arrow).
    \item Cut out regions (ITS1, ITS2, full ITS, LSU) using ITSx.
    \item Quality filter - up to 3 expected errors.
  \end{itemize}
\end{frame}

\begin{frame}{Expected error rate}
\framesubtitle{After \texttt{Arrow} and \texttt{ITSx}}
  
<<errorrate, fig.width = 12, cache.check = reads_subset>>=
filter(reads_subset, tech == "PacBio") %>%
  ggplot(aes(x = eexp,
             group = paste(dataset, region),
             color = dataset)) +
  geom_line(aes(y = 1 - ..y..), stat = "ecdf", size = 1) +
  scale_x_continuous(name = "Expected number of errors",
                     trans = reverselog_trans(10),
                     limits = c(NA, 1e-5),
                     oob = squish) +
  facet_wrap(~region, ncol = 2, strip.position = "right") +
  scale_y_continuous(name = "Fraction passing") +
  # scale_color_viridis_d(option = "C") +
  geom_vline(aes(xintercept = 3), linetype = 2)
@
\end{frame}

\begin{frame}{Bioinformatics}
\framesubtitle{continued...}
  \begin{itemize}
    \item Find ITS2 amplicon sequence variants (ASVs) with \texttt{dada2}.
    \item Generate LSU consensus for each ASV
  \end{itemize}
\end{frame}

\begin{frame}{Amplicon and individual region lengths}
\framesubtitle{After \texttt{Arrow} and \texttt{ITSx}}
  
<<length, fig.width = 12, cache.check = reads_subset>>=
filter(reads_subset, tech == "PacBio", eexp <= 3) %>%
  ggplot(aes(x = length,
             group = paste(dataset, region),
             color = dataset, fill = dataset)) +
  geom_area(stat = "density", size = 1, alpha = 0.5) +
  scale_x_log10(name = "Length (bp)",
                limits = c(50, NA),
                oob = squish) +
  facet_wrap(~region, ncol = 2, strip.position = "right", scales = "free_y") +
  scale_y_continuous(name = "Fraction passing") +
  # scale_color_viridis_d(option = "C") +
  geom_vline(aes(xintercept = 3), linetype = 2)
@
\end{frame}

\begin{frame}{Long vs. short amplicons}
\framesubtitle{ASV read counts}
\begin{overprint}
\onslide<1>
Total ASVs:
\begin{itemize}
  \item Long amplicon: 646 ASVs
  \item Short amplicon: 531 ASVs
\end{itemize}

\onslide<2>
<<readplot, fig.width = 12, message=FALSE>>=
loadd(big_seq_table_ITS2)
asv_table <- big_seq_table_ITS2 %>%
  tibble::as_tibble(rownames = "filename") %>%
  tidyr::gather(key = "seq", value = "reads", -1) %>%
  dplyr::filter(reads >= 1) %>%
  tidyr::extract(col = "filename",
                 into = c("seq_run", "plate", "well", "dir", "region"),
                 regex = "([a-z]+_\\d+)_(\\d+)-([A-H]1?[0-9])([fr]?)-([:alnum:]+).+") %>%
  dplyr::filter(seq_run %in% c("pb_483", "pb_500")) %>%
  dplyr::group_by(seq_run, seq) %>%
  dplyr::summarize(reads = sum(reads)) %>%
  tidyr::spread(key = seq_run, value = reads, fill = 0) %>%
  dplyr::rename(short_pacbio = pb_483, long_pacbio = pb_500)

asv_table %>% 
  filter(short_pacbio > 0, long_pacbio > 0) %>%
  ggplot(aes(x = short_pacbio, y = long_pacbio)) +
  geom_point(alpha = 0.2) +
  scale_x_log10() +
  scale_y_log10() +
  xlab("Short amplicon reads") +
  ylab("Long amplicon reads") +
  geom_rug(aes(y = long_pacbio, x = 1), sides = "l", alpha = 0.2,
           data = asv_table %>% filter(short_pacbio == 0)) +
  geom_rug(aes(y = short_pacbio), sides = "b", alpha = 0.2,
           data = asv_table %>% filter(long_pacbio == 0)) +
  stat_smooth(method = "loess", formula = y ~ x) +
  geom_abline(aes(intercept = mean(log10(long_pacbio[long_pacbio >= 1])) -
                    mean(log10(short_pacbio[short_pacbio >= 1])), slope = 1),
              linetype = 2L, alpha = 0.5)
@
\end{overprint}
\end{frame}

% \begin{frame}{Tree}
% \framesubtitle{Kingdoms}
% <<kingtree, fig.width = 12, message=FALSE>>=
% seqhash <- function(seq, len = 8) UseMethod("seqhash")
% seqhash.character <- function(seq, len = 8) {
%   h <- purrr::map_chr(seq, digest::digest)
%   stringr::str_sub(h, end = len)
% }
% seqhash.XStringSet <- function(seq, len = 8) {
%   seqhash.character(as.character(seq))
% }
% treetax <- taxon_LSUcons_rdp %>%
%   mutate_at("seq", seqhash) %>%
%   select(-Taxonomy, -nreads) 
% outgroups <- treetax %>% filter(!is.na(Kingdom), Kingdom %in% c("Alveolata", "Chromista", "Plantae"), seq %in% longtree$tip.label)
% p <- ggtree(tr = longtree)
% longtree <- reroot(longtree, MRCA(p, outgroups$seq))
% 
% ggtree(groupOTU(longtree, split(treetax$seq, treetax$Kingdom))) +
%   aes(color = group) + scale_color_brewer(type = "qual", palette = "Paired") +
%   theme(legend.position = NULL)
% @
% \end{frame}
% 
% \begin{frame}{Tree}
% \framesubtitle{Phyla}
% <<phytree, fig.width = 12, message=FALSE>>=
% ggtree(groupOTU(longtree, split(treetax$seq, treetax$Phylum))) +
%   aes(color = group) + scale_color_brewer(type = "qual", palette = "Paired") +
%   theme(legend.position = NULL)
% @
% \end{frame}
% 
% \begin{frame}{Tree}
% \framesubtitle{Families --- Basidiomycota}
% <<basidtree, fig.width = 12, message=FALSE>>=
% basids <- treetax %>% filter(!is.na(Class),
%                              Class %in% c("Geminibasidiomycetes",
%                                          "Tremellomycetes",
%                                          "Exobasidiomycetes",
%                                          "Microbotryomycetes"),
%                              seq %in% longtree$tip.label)
% 
% basidtree <- phylobase::phylo4(longtree) %>%
%   phylobase::subset(mrca = basids$seq) %>%
%   as("phylo")
% 
% basids2 <- treetax %>% filter(seq %in% basidtree$tip.label)
% 
% ggtree(groupOTU(basidtree, split(basids2$seq, basids2$Order)), layout = "circular") +
%   aes(color = group) +
%   scale_color_brewer(type = "qual", palette = "Paired", limits = unique(na.omit(basids2$Order))) +
%   theme(legend.position = NULL)
% @
% \end{frame}

\begin{frame}{Correlogram}

<<correlogram, fig.width = 12>>=
bind_rows(as.data.frame(bc_corr_long$mantel.res) %>%
            mutate(tdist = "same", seq_run = "long"),
          as.data.frame(bc_corr_long2$mantel.res) %>%
            mutate(tdist = "diff", seq_run = "long"),
          as.data.frame(bc_corr_short$mantel.res) %>%
            mutate(tdist = "same", seq_run = "short"),
          as.data.frame(bc_corr_short2$mantel.res) %>%
            mutate(tdist = "diff", seq_run = "short")) %>%
  filter(complete.cases(.)) %>%
  mutate(Significant = `Pr(corrected)` < 0.05) %>%
  ggplot(aes(class.index, Mantel.cor, group = tdist, color = tdist, shape = Significant)) +
  geom_point(size = 3) +
  geom_line(alpha = 0.3, size = 1.5) +
  geom_hline(yintercept = 0, linetype = 2, alpha = 0.5) +
  # geom_line(aes(y = exp(predict(bc_fit)))) +
  scale_y_continuous(limits = c(-0.05, NA), name = "Spatial Distance correlation") +
  scale_color_discrete(name = "Year") +
  xlab("distance (m)") +
  facet_wrap("seq_run") +
  scale_shape_manual(values = c("TRUE" = 16, "FALSE" = 1))
@
\end{frame}

\begin{frame}{Acknowledgements}
\begin{columns}[t]
  \begin{column}{0.5\linewidth}
    \textbf{Uppsala University}\\
    Anneli Svanholm\\
    Bobby Sulistyo\\
    Ylva Strid\\
    Kerri Kluting
  \end{column}
  \begin{column}{0.5\linewidth}
    
    \textbf{University of Parakou}\\
    Roel Houdanon\\
    Evans Codjia\\
    Souleymane Boni\\
    Hyppolite Aignon\\
    Sylvestre Badou\\
    Gerard Laourou\\
    Aziz Boukary\\
    Olivia Fayedi
  \end{column}
\end{columns}
\includegraphics[width=0.5\linewidth]{formas}
\end{frame}

\end{document}