---
title: "Comparison of ASVs and OTUs between PacBio and Ion Torrent"
author: "Brendan Furneaux"
date: "10 June 2019"
output:
  beamer_presentation:
    latex_engine: xelatex
---

```{r setup, include = FALSE}
library(magrittr)
library(tidyverse)
library(here)
library(glue)
library(Biostrings)
library(ShortRead)
library(ggplot2)
library("scales")
library(drake)

source(here("scripts/map_LSU.R"))
knitr::opts_chunk$set(echo = FALSE, device = "tikz")
theme_set(theme_bw(base_size = 24) + theme(legend.direction = "vertical"))
```

# Read number from PacBio vs. IonTorrent ASVs (ITS2)
```{r}
loadd(big_seq_table_ITS2)
asv_table <- big_seq_table_ITS2 %>%
  tibble::as_tibble(rownames = "filename") %>%
  tidyr::gather(key = "seq", value = "reads", -1) %>%
  dplyr::filter(reads >= 1) %>%
  tidyr::extract(col = "filename",
                 into = c("seq.run", "plate", "well", "dir", "region"),
                 regex = "([a-z]+_\\d+)_(\\d+)-([A-H]1?[0-9])([fr]?)-([:alnum:]+).+") %>%
  dplyr::filter(seq.run %in% c("is_057", "pb_500")) %>%
  dplyr::group_by(seq.run, seq) %>%
  dplyr::summarize(reads = sum(reads)) %>%
  tidyr::spread(key = seq.run, value = reads, fill = 0)

asv_table %>% 
  filter(is_057 > 0, pb_500 > 0) %>%
  ggplot(aes(x = is_057, y = pb_500)) +
  geom_point(alpha = 0.2) +
  scale_x_log10() +
  scale_y_log10() +
  xlab("Ion Torrent Reads") +
  ylab("PacBio Reads") +
  geom_rug(aes(y = pb_500, x = 1), sides = "l", alpha = 0.2,
           data = asv_table %>% filter(is_057 == 0)) +
  geom_rug(aes(y = is_057), sides = "b", alpha = 0.2,
           data = asv_table %>% filter(pb_500 == 0)) +
  stat_smooth(method = "loess", formula = y ~ x) +
  geom_abline(aes(intercept = mean(log10(pb_500[pb_500 >= 1])) -
                    mean(log10(is_057[is_057 >= 1])), slope = 1),
              linetype = 2L, alpha = 0.5)

```



# Read number from PacBio vs. IonTorrent ASV-OTUs (ITS2)

```{r}
otu <- read_tsv(here("data/clusters/ITS2.table"),
                col_types = "ciii") %>% select(is_057, pb_500) %>%
  filter(rowSums(.) > 0)
otu %>%
  filter(is_057 > 0, pb_500 > 0) %>%
  ggplot(aes(x = is_057, y = pb_500)) +
  geom_point(alpha = 0.2) +
  scale_x_log10() +
  scale_y_log10() +
  xlab("Ion Torrent Reads") +
  ylab("PacBio Reads") +
  geom_rug(aes(y = pb_500, x = 1), sides = "l", alpha = 0.2,
           data = otu %>% filter(is_057 == 0)) +
  geom_rug(aes(y = is_057), sides = "b", alpha = 0.2,
           data = otu %>% filter(pb_500 == 0)) +
  stat_smooth(method = "loess", formula = y ~ x) +
  geom_abline(aes(intercept = mean(log10(pb_500[pb_500 >= 1])) -
                    mean(log10(is_057[is_057 >= 1])), slope = 1),
              linetype = 2L, alpha = 0.5)
```

# IonTorrent ITS2 ASVs: coverage by long PacBio
```{r}
asv_table %>%
  arrange(desc(is_057)) %>%
  filter(is_057 > 0) %>%
  ggplot(aes(y = is_057, x = seq_along(is_057), fill = pb_500 > 0)) +
  geom_col(width = 1) +
  xlab("ASVs") +
  ylab("reads") +
  scale_fill_discrete(name = "PacBio") +
  scale_y_continuous(sec.axis = sec_axis(~ ./sum(.), name = "frac. of comm."))
```
`r asv_table %>% filter(is_057 > 0) %>% {formatC(sum(.$pb_500 > 0) / length(.$is_057)*100, digits = 1, flag = "0", format = "f")}`% of IonTorrent ASVs (`r asv_table %>% {formatC(sum(.$is_057[.$pb_500 > 0]) / sum(.$is_057) * 100, digits = 1, flag = "0", format = "f")}`% of reads) also include at least one PacBio read.

# IonTorrent ITS2 ASV-OTUs: coverage by long PacBio
```{r}
otu %>%
  arrange(desc(is_057)) %>%
  filter(is_057 > 0) %>%
  ggplot(aes(y = is_057, x = seq_along(is_057), fill = pb_500 > 0)) +
  geom_col(width = 1) +
  xlab("OTUs") +
  ylab("reads") +
  scale_fill_discrete(name = "PacBio") +
  scale_y_continuous(sec.axis = sec_axis(~ ./sum(.), name = "frac. of comm."))
```
`r otu %>% filter(is_057 > 0) %>% {formatC(sum(.$pb_500 > 0) / length(.$is_057)*100, digits = 1, flag = "0", format = "f")}`% of IonTorrent OTUs (`r otu %>% {formatC(sum(.$is_057[.$pb_500 > 0]) / sum(.$is_057) * 100, digits = 1, flag = "0", format = "f")}`% of reads) also include at least one PacBio ASV.

# PacBio ITS2 ASVs: coverage by Ion Torrent
```{r}
asv_table %>%
  arrange(desc(pb_500)) %>%
  filter(pb_500 > 0) %>%
  ggplot(aes(y = pb_500, x = seq_along(pb_500), fill = is_057 > 0)) +
  geom_col(width = 1) +
  xlab("ASVs") +
  ylab("reads") +
  scale_fill_discrete(name = "Ion Torrent") +
  scale_y_continuous(sec.axis = sec_axis(~ ./sum(.), name = "frac. of comm."))
```
`r asv_table %>% filter(pb_500 > 0) %>% {formatC(sum(.$is_057 > 0) / length(.$pb_500)*100, digits = 1, flag = "0", format = "f")}`% of PacBio ASVs (`r asv_table %>% {formatC(sum(.$pb_500[.$is_057 > 0]) / sum(.$pb_500) * 100, digits = 1, flag = "0", format = "f")}`% of reads) also include at least one IonTorrent read

# PacBio ITS2 ASV-OTUs: coverage by Ion Torrent
```{r}
otu %>%
  arrange(desc(pb_500)) %>%
  filter(pb_500 > 0) %>%
  ggplot(aes(y = pb_500, x = seq_along(pb_500), fill = is_057 > 0)) +
  geom_col(width = 1) +
  xlab("OTUs") +
  ylab("reads") +
  scale_fill_discrete(name = "Ion Torrent") +
  scale_y_continuous(sec.axis = sec_axis(~ ./sum(.), name = "frac. of comm."))
```
`r otu %>% filter(pb_500 > 0) %>% {formatC(sum(.$is_057 > 0) / length(.$pb_500)*100, digits = 1, flag = "0", format = "f")}`% of PacBio OTUs (`r otu %>% {formatC(sum(.$pb_500[.$is_057 > 0]) / sum(.$pb_500) * 100, digits = 1, flag = "0", format = "f")}`% of reads) also include at least one IonTorrent ASV.

# ASV sample coverage
```{r}
asv_samps <- big_seq_table_ITS2 %>%
  tibble::as_tibble(rownames = "filename") %>%
  tidyr::gather(key = "seq", value = "reads", -1) %>%
  tidyr::extract(col = "filename",
                 into = c("seq.run", "plate", "well", "dir", "region"),
                 regex = "([a-z]+_\\d+)_(\\d+)-([A-H]1?[0-9])([fr]?)-([:alnum:]+).+") %>%
  dplyr::filter(reads >= 1, seq.run %in% c("pb_500", "is_057")) %>%
  dplyr::group_by(seq.run, well, seq) %>%
  dplyr::summarize(reads = sum(reads)) %>%
  dplyr::ungroup()

asv_samp_plot <- asv_samps %>%
  dplyr::group_by(seq.run, seq) %>%
  dplyr::summarize(sample = n_distinct(well)) %>%
  tidyr::spread(key = seq.run, value = sample, fill = 0)

asv_samp_plot %>%
  filter(is_057 > 0, pb_500 > 0) %>%
  ggplot(aes(x = is_057, y = pb_500)) +
  geom_point(alpha = 0.2) +
  scale_x_log10() +
  scale_y_log10() +
  xlab("Ion Torrent wells") +
  ylab("PacBio wells") +
  geom_rug(aes(y = pb_500, x = 1), sides = "l", alpha = 0.2,
           data = asv_samp_plot %>% filter(is_057 == 0)) +
  geom_rug(aes(y = is_057), sides = "b", alpha = 0.2,
           data = asv_samp_plot %>% filter(pb_500 == 0)) +
  stat_smooth(method = "loess", formula = y ~ x) +
  geom_abline(aes(intercept = mean(log10(pb_500)) -
                    mean(log10(is_057)), slope = 1),
              linetype = 2L, alpha = 0.5)
  
  
```

# OTU Sample coverage

```{r}
otumap = read_tsv(here("data/clusters/ITS2.uc"),
                  col_names = c("Type", "ID", "Length", "Identity", "Strand",
                                "X1", "X2", "Align", "Query", "Target"),
                  col_types = "cccccccccc") %>%
  dplyr::select(otu = ID, seq = Query) %>%
  dplyr::mutate_at("seq", stringr::str_remove_all, ";.+") %>%
  unique()

otu_samps <- asv_samps %>%
  dplyr::mutate_at("seq", seqhash) %>%
  dplyr::left_join(otumap, by = "seq")

otu_samp_plot <- otu_samps %>%
  dplyr::group_by(seq.run, otu) %>%
  dplyr::summarize(sample = n_distinct(well)) %>%
  tidyr::spread(key = seq.run, value = sample, fill = 0)

otu_samp_plot %>%
  filter(is_057 > 0, pb_500 > 0) %>%
  ggplot(aes(x = is_057, y = pb_500)) +
  geom_point(alpha = 0.2) +
  scale_x_log10() +
  scale_y_log10() +
  xlab("Ion Torrent wells") +
  ylab("PacBio wells") +
  geom_rug(aes(y = pb_500, x = 1), sides = "l", alpha = 0.2,
           data = otu_samp_plot %>% filter(is_057 == 0)) +
  geom_rug(aes(y = is_057), sides = "b", alpha = 0.2,
           data = otu_samp_plot %>% filter(pb_500 == 0)) +
  stat_smooth(method = "loess", formula = y ~ x) +
  geom_abline(aes(intercept = mean(log10(pb_500)) -
                    mean(log10(is_057)), slope = 1),
              linetype = 2L, alpha = 0.5)
```

# Costs going forward

 - SciLife still has the short amplicon libraries for Ion Torrent
 - Alternative 1: So we can resequence them on two Ion Torent S5 530 chips and demultiplex
   - Cost (my best guess) $2 \text{ runs} \times 14000 \frac{\text{sek}}{\text{run}} = 28000 \text{ sek}$
 - Alternative 2: Pool these two libraries (total ~150 samples) with ITS2 library from spatial paper (same primers, ~100 samples), sequence on one Illumina MiSeq 2Ã—300bp lane (~18M read pairs)
   - Compare total community recovered from PacBio, Ion Torrent, and Illumina on the transect samples
   - Cost (best guess) $3 \text{ libraries} \times 1000 \frac{\text{sek}}{\text{ library}} + 1 \text{ run} \times 17000 \frac{\text{sek}}{\text{run}} = 20000 \text{ sek}$
 - Alternative 3: Forget the libraries from the transect paper and sequence a long amplicon library for the spatial paper on Sequel
    - We currently have 18 pools of 5 samples each. But not demultiplexed yet!
    - Cost - ask Anna about Peter's sequencing cost.
