---
title: "Long- and short-read metabarcoding technologies reveal similar spatio-temporal structures in fungal communities"
author: 
  - Brendan Furneaux^1,\*^,
  - Mohammad Bahram^2,3^,
  - Anna Rosling^4^,
  - Nourou S. Yorou^5^,
  - Martin Ryberg^1^
date: |
      | ^1^Program in Systematic Biology,
        Department of Organismal Biology,
        Uppsala University,
        Uppsala, Sweden
      | ^2^Department of Ecology,
        Swedish University of Agricultural Sciences,
        Uppsala, Sweden
      | ^3^Institute of Ecology and Earth Sciences,
        University of Tartu,
        Tartu, Estonia
      | ^4^Program in Evolutionary Biology,
        Department of Ecology and Genetics,
        Uppsala University,
        Uppsala, Sweden
      | ^5^Research Unit in Tropical Mycology and Plant-Fungi Interactions,
        LEB,
        University of Parakou,
        Parakou, Benin
      | ^\*^Corresponding author, `brendan.furneaux@ebc.uu.se`
bibliography:
  - "all.bib"
  - "R.bib"
biblio-style: apa
always_allow_html: true
output:
  bookdown::pdf_document2:
    template: bare_template.tex
    fig_crop: true
    keep_tex: true
    toc: false
    latex_engine: lualatex
    includes:
      in_header: preamble.tex
    citation_package: biblatex
    suppress-bibliography: true
  bookdown::word_document2: 
    keep_md: yes
fontsize: 12pt
geometry: ["left=2.5cm", "right=2.5cm", "top=3cm", "bottom=3cm"]
graphics: yes
---

```{r setup, include=FALSE}
library("knitr")
library("tikzDevice")
library("ggplot2")
library("magrittr")
library("tidyverse")
library("drake")
library("here")
library("phyloseq")
tryCatch(
  stopifnot(dir.exists(tinytex::tinytex_root())),
  error = function(e) tinytex::install_tinytex()
  )

tinytex::tlmgr_install("pdfcrop")

knitr::opts_chunk$set(
  dev = "cairo_pdf",
  # dev.args = list(pointsize=32),
  echo = FALSE,
  # results = 'asis',
  # fig.show = 'asis',
  fig.width = 6,
  fig.height = 4.5,
  message = FALSE,
  warning = FALSE,
  external = TRUE,
  out.width = "70%",
  fig.align = "center"
)

base.dir <- regmatches(getwd(), regexpr(".*oueme-fungi-transect", getwd()))
write_bib(c("base", "phyloseq", "vegan", "metacoder"), file = "R.bib", tweak = TRUE)

if (!exists("percent", mode = "function")) source(here::here("scripts", "output_functions.R"))

light_plan <- readRDS(here("data/plan/drake_light.rds"))
theme_set(theme_bw())

datasets <- read_csv(here("config/datasets.csv"), col_types = "cccccccicccccccccccicc")
regions <- read_csv(here("config/regions.csv"), col_types = "cccciiiiic")
```

<!-- R1: This study seeks to determine how both sequencing target length and technology affect the spatial and temporal turnover of soil fungal communities in an African forest. The authors take an ambitious approach to these research questions by comparing multiple short-read technologies (Illumina, PacBio, and Ion Torrent) to a related long-read technology (PacBio), allowing them to make many valuable comparisons across datasets. Along with field sample replication, both spatially and temporally, the authors carefully consider a number of different bioinformatic steps for each sample, e.g. among taxonomic databases as well as whether to cluster ASV data or not. Importantly, after discovering that long read data has inherent properties that make it poorly suited for the popular DADA2 bioinformatic pipeline, the authors generate a series of new R-based tools to analyze long read data in ways that greater increase data retention. Further, they use a clever set of covariance models to improve the taxonomic assignments of ASVs that are not assigned at the bulk taxonomy step. This additional step is important in increasing the amount of sequence data that can be annotated to putative functional guilds, which in this case focuses on ECM fungi.  

Overall, I found the study to be a really excellent demonstration of how different sequencing technologies and target lengths can be compared for fungi. The overall conclusion that the ecological outcome is largely similar between the different datasets is heartening. I am quite sure that longer reads will continue to increase in popularity as costs decline and this study represents an important bridge in that transition – both for its careful benchmarking but also for its presentation of tools that allow for better incorporation of long read sequence data into existing bioinformatic pipelines. I hope that my comments below help the authors make some relatively minor improvements, particular in adding more biological interpretation in a few places to help more deeply understand the patterns they have identified. -->

<!-- R2: The authors have collected an interesting data set but I feel strongly that the work requires significant improvement to the Introduction (particularly the justification) and Discussion (particularly some explanations, even speculative) sections, as detailed below in my specific comments.

Regarding the justification and conclusions, a quick scan of the literature revealed a study that used long(er) reads (covering SSU, ITS, LSU; Heeger et al. 2019 in this journal) and also separated these prior to clustering and assignment. I therefore see the contribution as stated here (line 614) as overstated except that denoising was applied prior to clustering. Phylogenetic reconstruction was also used here, although the conclusions are not justified (see my comment below). I see an important contribution of this work as developing a new method to separate fragments; however, although it should be clear (i.e. examined in the Introduction) why existing approaches like ITSx or others are not suitable. 

"Long-read" has become a vague descriptor and should be clearly defined early on for this manuscript (in the abstract). How many bp or markers? I have seen the term used in metabarcoding to mean anything from a few bp longer than a MiSeq PE 2x250, to nearly 5000 bp covering multiple rRNA genes.-->

<!-- R3: This paper analyzes a transect of soils in Africa and reports how results differ from 3 differing sequencing technologies and variations to bioinformatics platforms. With sequencing technologies changing so rapidly, and so many methods to choose from, I’m sure this paper would be of interest to the Molecular Ecology Resources audience, as it provides useful data comparing sequencing technologies. I congratulate the author on herculean efforts to write so many R packages, for making his packages and scripts publicly available, and for making a pipeline that allegedly improves ASV recovery for long amplicons. I personally would find the paper more interesting if a treatment other than spatial turnover had been applied to these samples, like before and after a disturbance, or a different host tree, and one could examine if the treatment of interest gave similar inferences with different sequencing technologies. However, I believe their claim that sequencing technologies yield similar results. I do think the paper would be improved by adding in some alpha diversity analyses, which are likely to be less robust to vagaries of sequencing technologies than beta diversity analyses. I also did not see any discussion of normalization for variation in number of sequences per sample, which needs to be accounted for. Instead of figure 2 (which can be moved to the supplements), I would prefer instead to see a figure about correlations of alpha diversity among methods. I also think species accumulation curves would be useful to see. I have also suggested several ways the writing could be slightly altered to make it clearer and more engaging for the reader. The results section is quite dry, reporting the statistics rather than reporting the interesting results and using the statistics to support it. For example, “Microbial communities are spatially autocorrelated (R2 = 0.90, p <0.05)” is more interesting to read than the “Bray-curtis dissimilarity showed…” If some of the results section could be re-written to focus on the results rather than reporting statistics, it would be more engaging. I also think the introduction needs to be edited to be more cohesive with the rest of the document. For example, only abbreviations were used in the introduction but they were fully spelled out at numerous points in the methods, and this made the introduction unnecessarily difficult to follow. I also found that the discussion repeated the results a lot and thus could be truncated or re-focused to set the results more in the context of the current literature (for example some citations of similar papers was missing). Finally, I will freely admit that some portions of the methods section, like the discussion of LSUx, went over my head and were very difficult for me to understand and/or follow. I’m not sure if someone more bioinformatically inclined could do a better job reviewing some of the more technical portions or if they just need to be written more clearly. I actually spent 3x the amount of time I usually devote to reviews on this paper, and I made a lot of efforts to provide constructive criticisms of how this paper could be improved to be clearer for the reader, so I hope that they are helpful. See specific line comments below.-->

# Abstract {-}

Fungi form diverse communities and play essential roles in many terrestrial ecosystems, yet there are methodological challenges in taxonomic and phylogenetic placement of fungi from environmental sequences.
To address such challenges we investigated spatio-temporal structure of a fungal community using soil metabarcoding with four different sequencing strategies: short amplicon sequencing of the ITS2 region (300--400\ bp) with Illumina MiSeq, Ion Torrent Ion S5, and PacBio RS II, all from the same PCR library, as well as long amplicon sequencing of the full ITS and partial LSU regions (1200--1600\ bp) with PacBio RS II.
Resulting community structure and diversity depended more on statistical method than sequencing technology.
The use of long-amplicon sequencing enables construction of a phylogenetic tree from metabarcoding reads, which facilitates taxonomic identification of sequences.
However, long reads present issues for denoising algorithms in diverse communities.
We present a solution that splits the reads into shorter homologous regions prior to denoising, and then reconstructs the full denoised reads.
In the choice between short and long amplicons, we suggest a hybrid approach using short amplicons for sampling breadth and depth, and long amplicons to characterize the local species pool for improved identification and phylogenetic analyses.

# Introduction

Fungi are key drivers of nutrient cycling in terrestrial ecosystems.
One important guild of fungi form ectomycorrhizas (ECM), a symbiosis between fungi and plants in which fungal hyphae enclose the plant's fine root tips.
The fungi provide nutrients and protection from pathogens in exchange for carbon from the plant [@Smith2010].
Approximately 8% of described fungal species are thought to take part in ECM symbiosis [@ainsworth2008ainsworth;@Rinaldi2008].
Although only about 2% of land plant species form ECM, these include ecologically and economically important stand-forming trees belonging to both temperate and boreal groups such as Pinaceae and Fagaceae, and tropical groups such as Dipterocarpaceae, *Uapaca* (Phyllanthaceae) and Fabaceae tr. Amherstieae [@brundrett2017a], which together represent approximately 60% of tree stems globally [@steidinger2019].

Although ECM fungi form many well-known mushrooms (e.g., *Amanita*, *Cantharellus*, *Boletus*), some instead produce inconspicuous (e.g., *Tomentella*) or no (e.g., *Cenococcum*) fruit bodies.
Even when fruitbodies are large, they are ephemeral, so study of ECM communities is facilitated by looking at vegetative structures [@horton2001].
Unlike many saprotrophic fungi which grow easily in axenic culture, ECM fungi are usually difficult to culture, so DNA barcoding is increasingly used to investigate vegetative structures in the field.
The advent of high-throughput sequencing (HTS) has facilitated such studies by providing enough sequencing depth for metabarcoding of bulk environmental samples such as soils [@Lindahl2013].

As additional techniques and methods are developed for HTS, there is an increasing array of choices for researchers investigating fungal communities.
Fungal metabarcoding studies using short-read HTS technologies such as 454 Pyrosequencing, Illumina, and Ion Torrent
have usually targeted the rDNA internal transcribed spacer regions ITS1 or ITS2,
which are the standard molecular barcode for fungi, providing sufficient resolution to distinguish fungal species in many groups, and which are usually short enough for HTS [@schoch2012;@Lindahl2013].
In some groups such as arbuscular mycorrhizal fungi, variable regions of the rDNA small subunit (SSU) are the barcode of choice [@opik2010], and variable regions of the rDNA large subunit (LSU) have also been used for barcoding [@house2016;@tedersoo2015].
The resulting sequencing reads are clustered by sequence similarity to form operational taxonomic units (OTUs), which are then used as the units for further community analysis [@Lindahl2013].
If taxonomic identification is desired in order to put OTUs in a wider context and associate functional information, it has usually been performed by database searches using BLAST [@altschul1990;@Lindahl2013] with public databases such as GenBank [@benson2013] and Unite [@nilsson2019a].
However, this approach comes with some potential weaknesses, discussed below.

Different sequencing technologies have different capabilities in terms of sequencing depth and read length, as well as differing quality profiles and potential biases [@yang2013].
The rapid development of new HTS technologies, as well as subsequent iterative improvements in sequencing chemistry and read capacity, means that the technologies used in metabarcoding studies, along with any associated biases, change rapidly.
As an example, the first study using HTS metabarcoding of soil fungi was published in 2009 [@buee2009] using 454 Pyrosequencing; production of 454 sequencers was subsequently discontinued in 2015, and sales of reagents stopped in 2016 [@hollmer2013].
This brings into question the comparability of studies conducted only a few years apart.

ITS1 and ITS2 often have suitable variation to distinguish species, although closely related species may share identical ITS sequences in certain groups such as various Pezizomycotina [@schoch2012],
but this variablity means that they cannot be reliably aligned over the fungal kingdom [@Lindahl2013;@Tedersoo2018].
<!-- R2: This is true, but the fungi are a large and diverse group and thus it is no surprise that a highly variable rRNA marker can not be aligned. Perhaps this manuscript could discuss how meaningful a phylogenetic approach to analysis of complete fungal communities is anyway. Perhaps building a tree for each phylum is possible (and a realistic evolutionary model could be used). I think that this could be a contribution here - to explore how this variation could be dealt with using some knowledge of the clade "fungi". (Partly addressed?) -->
Additionally, the wide range of length variation of these regions may introduce bias in recovery of different taxa [@ihrmark2012;@tedersoo2015;@palmer2018].
Further bias is introduced by variation in the 5.8S region which separates the two ITS regions, as well as in the 5' end of LSU, which makes it difficult to design primers that are suitable for all fungi [@tedersoo2015].

Distance-based clustering conflates intra-species variation and sequencing error, and results are dataset-specific.
In contrast, more recent denoising methods such as DADA2 [@callahan2017], Deblur [@amir2017], and UNOISE2 [@edgar2016] utilize read quality information to control for sequencing error while preserving intra-species variation.
The resulting units are known as amplicon sequence variants (ASVs) or exact sequence variants (ESVs), as they should represent true amplicon sequences from the sample.
Unlike cluster-based OTUs, ASVs can capture variation of as little as one base pair, and have been suggested to be less dataset specific [@callahan2017].
Support for PacBio has recently been added to DADA2 [@callahan2019], but its application requires greater sequencing depth for longer reads, especially in high diversity samples.

Because both OTU clustering and denoised ASVs may "clump" different species into a single unit and "split" a single species into multiple units [@ryberg2015], diversity measures based on counting species within a community or shared species between two communities may give different results depending on the clustering threshold.
In contrast, phylogenetic community distance measures [@wong2016] are relatively insensitive to species/OTU delimitation, but require a phylogenetic tree.
Phylogenetic placement algorithms have been developed to place short amplicon reads onto a reference tree [@matsen2010; @berger2011], but are not easy to apply to ITS sequences because they require that the query sequences be aligned to a reference alignment.
Additionally, methods exist to place OTUs on a simplified tree based on taxonomic assignments [@tedersoo2018], or to create hybrid trees using ITS and a more conserved marker such as SSU or LSU based on matching taxonomic annotations in reference databases [@fouquier2016].

Assignment of taxonomic identities to environmental sequences is dependent on both the reference database and the algorithm used.
Although the public INSDC databases are often used for sequence identification, the open nature of submission to these databases results in a substantial fraction of incorrect taxonomic annotations [@nilsson2006;@steinegger2020] as well as sequences of poor technical quality [@nilsson2012].
Consequently taxonomic assignments based on these databases may be incorrect or inconsistent.
Several curated databases also exist which attempt to address these issues and which cover the whole fungal kingdom.
The Unite database is a more curated attempt to include all publically available ITS sequences (800\ 000 as of release 8.0), originally targetting fungi but now expanded to include all eukaryotes [@nilsson2019a], where efforts have been made to correct incorrect annotations and exclude low-quality sequences [@abarenkov2018a].
The Ribosomal Data Project [RDP, @cole2014] hosts two manually curated fungal barcode sequence databases, which are specifically intended for use in taxonomic assignment of sequences.
The Warcup ITS database contains 18\ 000 manually curated fungal ITS sequences [@deshpande2016], and the RDP fungal LSU training set contains 8000 manually curated LSU sequences from fungi and 3000 from other eukaryotic groups [@liu2012].
Although the quality of sequences and taxonomic annotations is undoubtedly higher in these more curated databases, they are inherently limited in taxonomic coverage and do not include the most recently published sequences.

Assigning taxonomy to unknown sequences using BLAST requires *a priori* choice of similarity thresholds for different taxonomic ranks.
Several algorithms specifically designed for taxonomic assignment have been published which instead use information about variability within different taxa in the reference database to assign unknown sequences, along with confidence estimates for these assignments, including the RDP Classifier [RDPC, @wang2007], SINTAX [@edgar2016a] and IDTAXA [@murali2018a] among others.
In addition, methods have been published which integrate predictions from multiple algorithms to increase the reliability of assignments [@somervuo2016;@gdanetz2017;@palmer2018].
However, all sequence-similarity based approaches are dependent on high taxonomic coverage in the reference database, making the placement of novel or undersampled groups problematic [@nilsson2016top;@Tedersoo2018].

Recent long-read HTS technologies such as Pacific Biosciences Single Molecule Real Time sequencing (PacBio) enable sequencing longer amplicons which include both the ITS regions and the flanking, more highly conserved SSU and/or LSU regions.
This can improve taxonomic placement of sequences that lack close database matches and allow the alignment of metabarcoding reads for subsequent phylogenetic analysis [@Tedersoo2018].
Information from phylogenetic trees produced from long-amplicon metabarcoding has the potential to both improve taxonomic assignment and provide alternative measures of community alpha and beta diversity.
However, long-read technologies are currently more expensive per read compared to short-read sequencing, and so their use entails a trade-off with sequencing depth and/or sample number [@Kennedy2018].

Because of the variety of sequencing platforms and analytical pipelines which have been used in metabarcoding studies, comparisons between studies may be difficult.
Here we investigated the effects of different sequencing strategies and post-analysis on biological conclusions using measurement of the spatiotemporal turnover rate of the fungal community in an ECM-dominated woodland in Benin by metabarcoding of bulk soil, sampled at narrow intervals, over two years.
We compare three different sequencing platforms (PacBio RS II, Illumina MiSeq, Ion Torrent Ion S5), long and short amplicons, three different taxonomic assignment algorithms (RDP classifier, SINTAX, IDTAXA)
and reference databases (Unite, Warcup, RDP),
and with non-phylogenetic and phylogenetic community distance measures.
We also present new algorithms for dividing the LSU into domains, combining denoising results from multiple domains as a strategy to capture more ASVs from long amplicons in diverse commmunities, and incorporating phylogenetic information into taxonomic assignments.

# Materials and Methods

## Sampling

Sampling was conducted at two sites (Ang: N 9.75456° E 2.14064°; Gan: N 9.75678° E 2.31058°) approximately 30\ km apart in the _Forêt Classée de l'Ouémé Supérieur_ (Upper Ouémé Forest Reserve) in central Benin.
Both sites were located in Soudanian savannah woodlands [@olson2001;@yorou2014] dominated by the ECM host tree _Isoberlinia doka_ (Caesalpinioideae).
At each site, 25\ soil samples were collected along a single 24 m linear transect at intervals of 1\ m in May 2015.
One third of the sample locations (3\ m spacing) were resampled one year later in June 2016.
For each sample, any coarse organic debris <!--Do you have any definition? Wider than 5 mm? MR-->was removed from the soil surface and a sample of approximately 5cm×5cm×5cm was extracted with a sterilized knife blade.
Each sample was sealed in a plastic zipper bag and homogenized by shaking and manually breaking apart soil aggregations.
Approximately 250\ mg total of soil was collected from two locations in the homogenized soil sample and placed into a separate 2.0\ mL microtube containing 750\ mL of field lysis/preservation buffer and lysis beads (Xpedition^TM^ Soil/Fecal DNA miniprep, Zymo Research Corporation, Irvine, California, USA).
Due to the inavailability of refrigeration, extraction was initiated in the field according to the manufacturers instructions by lysis using a handheld bead-beater (TerraLyser^TM^; Zymo Research Corporation).

An additional sample was collected at every sampling location (1-m spacing) in 2016 using LifeGuard^TM^ Soil Preservation Solution (MO BIO, Carlsbad, CA; USA) for preservation, without field lysis.
Sequencing results for these samples differed significantly (PERMANOVA with 9999 permutations, $p < `r readd(buffer_adonis)["buffer","p.value"] %>% formatC(format = "f")`$, $R^2 = `r readd(buffer_adonis)["buffer","R2"] %>% formatC(format = "f", digits = 2)`$) from samples preserved using the Xpedition^TM^ lysis buffer (Figures \@ref(fig:buffer-pcoa), \@ref(fig:buffer-compare-long), and \@ref(fig:buffer-compare-short)); as such the LifeGuard preserved samples were excluded from our spatial analyses.
However, reads from these samples were included in the full bioinformatics workflow, including ASV calling, OTU clustering, and phylogenetic trees.

## DNA extraction, amplification, and sequencing

After field lysis, DNA was extracted using the Xpedition^TM^ Soil/Fecal Prep kit (see above).
Samples preserved using LifeGuard were first centrifuged at 10000\ g for 1\ minute, after which the supernatant was removed and DNA was extracted from the remaining soil using the Soil/Fecal Prep kit as for the other samples.
DNA was quantified using fluorometrically using Quant-iT^TM^ PicoGreen^TM^ dsDNA (Thermo Fisher Scientific, Waltham, MA, USA) fluorescent indicator dye on a Infinite F200 plate spectrofluorometer (Tecan Trading AG, Männedorf, Switzerland) according to the manufacturer's protocol.

The DNA extracts were sequenced using four distinct strategies, with two different amplicon lengths (long and short; Figure\ \@ref(fig:regions)a) and three different technologies (PacBio, Ion Torrent, and Illumina) for the short amplicon.
Due to length restrictions for Ion Torrent and Illumina sequencing, long amplicons were only sequenced with PacBio.
The short amplicon (approximately 300\ bp) targeted the full ITS2 region as well as parts of the flanking 5.8S and large subunit (LSU) rDNA, using gITS7 [@ihrmark2012] as the forward primer and a mix of ITS4 [@white1990amplification] and ITS4a [@urbina2016] as the reverse primer (hereafter, ITS4m).
The long amplicon (approximately 1500\ bp) targeted the full ITS region including the 5.8S rDNA and approximately 950\ bp at the 5' end of the LSU, including the first three variable regions (Figure\ \@ref(fig:regions)a), using ITS1 [@white1990amplification] as the forward primer and LR5 [@vilgalys1990] as the reverse primer.
Each PCR run also included a blank sample and a positive control consisting of freshly extracted DNA from a commercially purchased fruitbody of *Agaricus bisporus*.

The forward primers for the short amplicon included sample-specific indexes and adapters for multiplexing (Supplementary File 1).
Amplification was performed by polymerase chain reaction (PCR) in 20µl reactions containing 200\ µM dNTP mix, 250\ µM indexed gITS7 primer,
150µM ITS4m, 2mM MgCl~2~, 0.1\ U *Taq* polymerase (Dream *Taq*, Thermo Fisher Scientific, Waltham, MA, USA) and 3--7\ ng purified DNA in Dream *Taq* buffer.
The reaction conditions were 10\ min at 95°, followed by 35\ cycles of 60\ s at 95°, 45\ s at 56°, and 50\ s at 72°, and finally 3\ min at 72°.
Each reaction was conducted in three technical replicates to reduce the effect of PCR stochasticity.

Both forward and reverse primers for the long amplicon included indexes for combinatorical multiplexing (Supplementary File 2).
PCR was performed as for the short amplicons, but with 500\ µM of each of the two primers.
Reaction conditions were 10\ min at 95°, 30\ cycles of 45\ s at 95°, 45\ s at 59°, and 90\ s at 72°, and finally 10\ min at 72°.
Each reaction was performed in three technical replicates as for short amplicons.

After pooling of technical replicates, amplicons were purified using SPRI beads [@vesterinen2016] and quantified fluorometrically as above.
An aliquot of 100\ ng of DNA from each sample (or the total PCR product if less than 100\ ng) was pooled into two libraries each for long and short amplicons.
Each library was sequenced using Single Molecule Real Time (SMRT) sequencing on a Pacific Biosciences (PacBio) RS II sequencer at the Uppsala Genome Center (UGC; Uppsala Genome Center, Science
for Life Laboratory, Dept. of Immunology, Genetics and Pathology, Uppsala University, BMC, Box 815, SE-752\ 37 UPPSALA, Sweden).
Short amplicon libraries were sequenced on two SMRT cells each, while long amplicon libraries were sequenced on four SMRT cells each.

Additionally, the same short amplicon PCR libraries were combined and sequenced using an Ion S5 (Ion Torrent) sequencer using one 520 chip at UGC, and a MiSeq (Illumina Inc.) sequencer using v3 chemistry with a paired-end read length of 300\ bp at the SNP&SEQ Technology Platform (Dept. of Medical Sciences, Uppsala University, BMC, Box 1432, SE-751\ 44 UPPSALA, Sweden) using one half of a lane.

## Bioinformatics

Circular consensus sequence (CCS) basecalls for PacBio sequences were made using `ccs` version 3.4 [@pacificbiosciences2019] using the default settings.
The resulting sequences, as well as the paired-end Illumina sequences, were demultiplexed and sequencing primers were removed using `cutadapt` version 2.8 [@martin2011].
Sequencing primers were similarly removed from the Ion Torrent sequences, but interference between the tagged gITS7 primers and the Ion XPress tags used in library prep made full demultiplexing of the Ion Torrent sequences impossible, and these reads were thus only analyzed as a pool.
For Ion Torrent and PacBio, reads were discarded if they did not have the appropriate primers on both ends.
Reads were searched in both directions, and reads where the primers were found in the reverse direction were reverse complemented before further analysis.
For Illumina sequences, read pairs were only retained when PCR primers were detected at the 5' ends of both the forward and reverse read.
Primers were also searched for and removed on the 3' ends of the reads, in case of readthrough with short amplicons.
Read pairs where the primers were found in reverse orientation were kept in separate files, but were retained in their original orientation until after denoising.

### Denoising and clustering

All amplicons were denoised using DADA2 version `r packageVersion("dada2")`a ccording to the ITS pipeline workflow [@callahan2016;@callahan2020], with technology-specific modifications for Ion Torrent [@callahan2020b] and PacBio [@callahan2019].
Although this was successful for the short amplicons on all technologies, only `r readd(region_table)$ASVs[readd(region_table)$region=="long"]` ASVs were obtained for the long amplicons, representing `r percent(readd(region_table)$reads[readd(region_table)$region == "long"]/as.integer(gsub(" ", "", readd(demuxcounts)["pb_500"])), format = "d")` of the trimmed reads.

We conclude that this poor performance was due to a combination of long amplicon length and low sequencing depth relative to community diversity.
The DADA2 algorithm requires that the seed sequence of each ASV be represented by at least two error-free reads [@callahan2016].
If sequencing errors are uniformly distributed, then the probability that a given read will be error-free is $(1-\epsilon)^L$, where $\epsilon$ is the sequencing error rate and $L$ is the read length in base pairs.
Then the number of reads of a given sequence that would be required to obtain two error-free reads in expectation is $2/(1-\epsilon)^L$.
For the combination of long reads (median $L=`r readd(demuxlength)['pb_500']` \text{ bp}$ after trimming) and moderate error rate (mean $\epsilon= `r format(readd(demuxqual)['pb_500'], digits = 4)`$ based on ccs quality scores) for the long amplicon in this study, the expected number of reads required to achieve two error-free reads is `r format(as.integer(round(2/(1 - readd(demuxqual)['pb_500'])^readd(demuxlength)['pb_500'])), big.mark = ",")`.
Given the high diversity relative to sequencing depth in this study (`r readd(region_table)$ASVs[readd(region_table)$seq_run == "pb_483" & readd(region_table)$region == "short"]` ASVs based on PacBio short amplicons, `r format(readd(demuxcounts)["pb_500"], format = "d", big.mark = ",")` trimmed long amplicon reads), this requirement could not have been met for the long amplicons except by the most abundant sequences.
In comparison, the equivalent requirement for the short amplicon ($L=`r readd(demuxlength)['pb_483']` \text{ bp}$, $\epsilon=`r readd(demuxqual)['pb_483']`$) is only `r round(2/(1 - readd(demuxqual)['pb_483'])^readd(demuxlength)['pb_483'], 1)` reads.

We therefore developed a new workflow to assemble ASVs from the long amplicons by splitting the reads into shorter homologous domains, independently denoising each segment, concatenating the denoised domains, and finally forming consensus sequences based on ITS2 identity.
We used this method, detailed below, for all of the PacBio and Ion Torrent datasets.
The method is not applicable to Illumina paired-end reads.

<!-- R3: honestly this section is above my head and I can’t judge its accuracy. Hopefully another review is more bioinformatically inclined and can judge it better. -->
Raw reads were divided into shorter regions by matching to covariance models (CM), which are similar to stochastic hidden markov models (HMM), but account for both nucleotide sequence and RNA secondary structure [@eddy1994].
First, the 5.8S rDNA was located in each read by searching for Rfam model RF0002 [@kalvari2018] using `cmsearch` from Infernal 1.1.2 [@nawrocki2013], and all bases before the 5.8S were assigned to ITS1.
No attempt was made to remove the approximately 12\ bp fragment of the SSU from the 5' end of ITS1 in the long amplicons; it was too short to be reliably detected by a CM or the HMMs employed by ITSx [@bengtsson-palme2013].
<!--This approach was able to identify both the 5' and 3' ends of the 5.8S rDNA more consistently and quickly than ITSx [@bengtsson-palme2013, data not shown].-->
A reference alignment including conserved RNA base pairing between and within the 5.8S and relevant portions of LSU was generated from the fungal 28S RNA seed alignment from RDP release 11.5 [@glockner2017; @cole2014] by truncating after the LR5 primer site and using the reference line to annotate the variable regions *sensu* @michot1984 and @raue1988.
A CM was generated from the alignment using `cmbuild` from Infernal.
The fragment of each read beginning with the 5.8S rDNA was then aligned to the CM using `cmalign` from Infernal.
The annotation line in the CM alignment for each read was then used to split the reads into alternating more-conserved and less-conserved domains as shown in Figure\ \@ref(fig:regions),
where LSU1-4 represent the conserved domains of LSU flanking the variable D1-3 domains [@michot1984].
For short amplicons, only (partial) 5.8S, ITS2, and (partial) LSU1 were extracted.
Code to extract the segments, including annotated seed alignments and CMs, is available in the new R package `LSUx`, available on Github at https://github.com/brendanf/LSUx.

Each of the extracted domains was independently filtered for length (Supplementary Table\ \@ref(tab:region-limits)) and a maximum of three expected errors, and denoised using DADA2.
To test the performance of DADA2 on intermediate-length amplicons, adjacent domains from each read were concatenated to form full ITS (ITS1 + 5.8S + ITS2), partial LSU (LSU1 + D1 + LSU2 + D2 + LSU3 + D3 + LSU4), and partial 32S (5.8S + ITS2 + LSU) datasets, and these were also denoised. 
The DADA2 error model was fit using the 5.8S region for long amplicons, and using the entire read for short amplicons.
Independent error models were fit for each sequencing strategy (i.e., long *vs.* short amplicons, different sequencing technologies).
DADA2 was run with complete pooling for PacBio libraries, and pseudo-pooling
for Ion Torrent libraries.
Chimeras within each region were removed using `removeBimeraDenovoTable` from DADA2.

For each raw read, the denoised sequences of the different domains were concatenated to form a set of full-length sequences.
For reads which were not assigned a denoised sequence for each domain, the raw read for the domain was used instead.
Because ITS2 is the most variable of the amplified domains as indicated by the greatest number of unique ASVs (Figure\ \@ref(fig:regions)b), reads with identical ITS2 regions are expected to have highly similar sequences in the other regions, unless the amplicon was chimeric.
Therefore, concatenated sequences with identical ITS2 were clustered, and each cluster was aligned in R using the `DECIPHER` package [@wright2015].
Outlier sequences, as determined by mean pairwise distance from the rest of the alignment, were removed from each alignment using the `odseq` package [@jehl2015], using the default threshold of 0.025.
The consensus of the remaining aligned sequences was assigned as the full-length ASV sequence for the entire cluster.
Full-length ASV sequences with more than three ambiguous bases (i.e., no nucleotide >50% at a given position) were removed.
The count and sample distribution of reads assigned to each full-length ASV were calculated in order to form a sample × ASV community matrix.
A similar process was used to generate a consensus ITS (ITS1--5.8S--ITS2) and LSU (LSU1--D1--LSU2--D2--LSU3--D3--LSU4) sequence for each ASV.
The process of assigning consensus full-length ASVs was carried out using the new `tzara` package for R, available at https://github.com/brendanf/tzara.

(ref:region-asv-s) rDNA regions

(ref:region-asv) **(a)** Partial map of rDNA showing the 5.8S rDNA, partial SSU and LSU rDNA, and internally transcribed spacer (ITS) regions.  D1-3 represent the first three variable regions in LSU, while LSU1-4 represent the conserved regions. Primer sites used in this study are indicated in red (forward primers) and blue (reverse primers), and the resulting amplicons are shown with green braces. **(b)** Total number of DADA2 ASVs vs. fraction of demultiplexed reads successfully mapped to ASVs for different rDNA regions extracted from a set of long PacBio amplicon sequences using `LSUx`. Shorter and more conserved regions yield a greater fraction of successfully mapped reads<!-- This sentence is a result shown in the figure (although it is also theoretically expected)-->. At a given fraction of mapped reads, more variable regions yield a greater number of unique ASVs<!-- This sentence is how to interpret the figure -->.

```{r regions, fig.cap="(ref:region-asv-s). (ref:region-asv)", fig.scap="(ref:region-asv-s)", fig.align = "center"}
ggpubr::ggarrange(
  readd(amplicon_plot),
  readd(length_diversity_plot),
  ncol = 1,
  labels = "auto",
  heights = c(0.6, 1)
)
```

Because the Illumina dataset consisted of paired-end reads, domains could not be extracted prior to denoising, and the ASVs generated according to the standard DADA2 workflow were used.
The ITS2 region was extracted from the ASVs using `LSUx` for comparison to the results from the other technologies.

For comparison with clustering-based methods, ASVs were clustered into operational taxonomic units (OTUs) at 97% similarity using VSEARCH v2.9.1 [@rognes2016].
<!-- R3: FYI, there is already a publication that compares OTU and ASV methods – Glassman and Martiny mSphere 2018 – and they result in the same inferences. Botnen et al Molecular Ecology Resources 2018 might also be worth citing.-->

### Phylogenetic inference and taxonomy assignment

Full length long amplicon ASVs were aligned using DECIPHER [@wright2015] with up to 10\ iterations of progressive alignment and conserved RNA secondary structure calculation and 10\ refinement iterations.
This alignment was truncated at a position after the D3 region corresponding to base 907 of the *Saccharomyces cerevisiae* S288C reference sequence for LSU, because several sequences had introns after this position, as also observed in several fungal species by @holst-jensen1999.
An ML tree was produced using RAxML version 8.2.12 [@stamatakis2014] using the GTR+GAMMA model and rapid bootstrapping with the MRE_IGN stopping criterion.

Taxonomic annotations of the RDP LSU fungal training set version 11.5 [@cole2014] and Warcup ITS training set [@deshpande2016] were mapped to the taxonomic classification system used in the Unite database version 8 [@nilsson2019a].
In particular, the classification for fungi was according to @tedersoo2018, and for non-fungal eukaryotes was according to the proposed system of @tedersoo2017c as described in [@tedersoodata2017].
Although the latter system is not formally published, it is consistent with the annotations for non-fungal eukaryotes in the Unite database.
Additionally, it is a system with both purportedly monophyletic taxa and a uniform set of taxon ranks, which make it more appropriate for sequence-based taxonomic assignment algorithms than more accepted classification systems such as that of the International Society of Protistologists [@adl2019], which utilizes hierarchical nameless ranks.

Preliminary taxonomic assignment was performed to genus level separately on the ITS region using Unite and Warcup and on the LSU region using RDP, respectively, as taxonomic references.
For each region/reference combination, taxonomy was assigned using three popular algorithms:
the RDP Naïve Bayesian Classifier [RDPC, @wang2007] as implemented in DADA2;
SINTAX [@edgar2016a] as implemented in VSEARCH v2.9.1 [@rognes2016];
and IDTAXA [@murali2018].
A relatively lax confidence threshhold of 50% was used for all three algorithms.
Each full-length ASV was thus given up to nine preliminary taxonomic assignments (three references $\times$ three algorithms).
ASVs from the short-amplicon datasets for which no matching long-amplicon ASV could be reconstructed were taxonomically assigned using Unite and Warcup on the full length of the short amplicon.
Conflicts between the assignments made by the different algorithms were later resolved using a phylogenetic method and a consensus method, which are described below.

The phylogenetic tree was rooted outside the kingdom Fungi by using the most abundant ASV which was confidently assigned to a non-fungal kingdom by all 6 applicable taxonomic assignment methods.
Assignments based on Warcup were not used at this step because non-Fungi are not included in the dataset.
The kingdom Fungi was identified as the minimal clade containing all ASVs which were confidently identified (consensus of at least 6 of 9 assignments) to a fungal phylum.
ASVs falling outside this clade were not included in downstream fungal community analysis.

For the ASVs from the long amplicon dataset, the preliminary taxonomic assignments were refined using the phylogenetic tree (Figure\ \@ref(fig:phylotax)).
Taxa at each rank were assigned to a node and all its descendants if that taxon was consistent with the reference-based taxonomic assignments for each of the descendants.
A taxon assignment was considered to be consistent if at least one algorithm assigned that taxon at greater than 50% confidence, or if no algorithm successfully classified the sequence at greater than 50% confidence.
The result of this process is twofold.
First, it gives a taxonomic assignment to ASVs which were previously unassigned if they are nested within a clade which is consistently assigned by the .
Second, it resolves conflicts between the different reference-based algorithms, but only if one of the alternatives is consistent with the assignments of other ASVs in the same clade.
This refinement algorithm is referred to as "PHYLOTAX", available in the new R package `phylotax` at https://github.com/brendanf/phylotax.

ASVs from the short amplicon datasets were refined using only the final strict consensus step, i.e., an assignment at a given rank was accepted if there was no conflict between the different assignment algorithms at greater than 50% confidence.
This refinement method is referred to as "Consensus".
Additionally, a hybrid method, was applied to the short amplicon datasets, in which assignments from PHYLOTAX were used for ASVs which could be linked by an identical ITS2 region to a long amplicon, and assignments from Consensus were used for the remaining ASVs.

## Effect of sequencing strategy on recovered community

The effect of sequencing technology and amplicon length on the recovered fungal community composition, as assessed by the Bray-Curtis dissimilarity, was determined by PERMANOVA [@anderson2003] using the `adonis2` function in the R package `vegan` [@R-vegan].
Comparisons were made between the fungal communities recovered by the three sequencing strategies that were successfully demultiplexed (Illumina short, PacBio short, PacBio long).
<!-- R3: I would say, the effect of the sequencing strategies and read length on the fungal community compositional dissimilarity, as assessed by either UNIFRAC or Bray-Curtis, was determined by PERMANOVA (cite Marti Anderson) using the Adonis function in vegan. I’m confused because you definitely mentioned in the intro UNIfrac but in this section only mention Bray-Curtis. Also you don’t mention any sort of normalization for differing sequence reads per sample, and that’s not really acceptable. I recommend avg.dist function in vegan. It lets you make the Bray-Curtis dissimilarity matrix subsampling for an even sequencing depth 100x and then take the median and perform a square root transformation all in one line of code.

Also, what no alpha diversity analysis? Beta-diversity analyses are generally pretty robust. I would be way more interested to see differences in alpha diversity by sequencing technology. Can you please report this (normalized for differences in sequencing depth)?
-->
In order to detect bias at larger taxonomic scales, ASVs were clustered according to the assigned taxonomic class.
Only samples where all three strategies yielded at least 100\ fungal reads (`r phyloseq::nsamples(readd(tech_class_table))/3` samples), and classes which represented at least 1% of reads in at least one sample (`r phyloseq::ntaxa(readd(tech_class_table))` classes), were included.
The community matrix was normalized to proportional abundance by dividing the counts for each class in each sample by the total number of reads for that sample, but were not rarefied to even sampling depth [@mcmurdie2014].
PERMANOVA included three terms: an indicator for soil sample, comprising all spatiotemporal effects; amplicon length (long vs. short); and sequencing technology (Illumina MiSeq vs. PacBio RS II).
Partial Principal Coordinates Analysis (PPCoA) was applied to the same dissimilarity matrix using the `capscale` function in `vegan` [@R-vegan].
Spatiotemporal effects were partialled out in order to visualize effects due to sequencing technology and amplicon length.

A similar analysis was also applied to only fungi classified as ECM, clustered at the family level.
Sequences were assigned as ECM based on taxonomic assignments using the FUNGuild database [@nguyen2016funguild] via the R package `FUNGuildR` (https://github.com/brendanf/FUNGuildR).
All taxa which included "Ectomycorrhiza" in the guild assignment at any level of confidence were included.

## Spatiotemporal analysis

Turnover scale is the distance at which two communities can be considered to be independent samples of the local species pool.
Knowledge of turnover scale is import when planning studies of local diversity and its environmental correlates.
It varies between different ecosystems and taxonomic groups.
Turnover scale is often measured by the range at which a Mantel correlogram indicates significant autocorrelation, or by fitting a function to an empirical distance-decay curve of community dissimilarity vs. distance [@legendre2012].

Ecological community dissimilarity matrices were calculated using the ASV/OTU based Bray-Curtis metric [@bray1957, for both long and short amplicons] and the phylogenetically based weighted UniFrac metric [@lozupone2005;@lozupone2007, for only long amplicons] in `phyloseq` version `r packageVersion("phyloseq")`.
Each of these distance matrices was used to calculate a Mantel correlogram with a 1\ m bin size for distances in the range of 0--12\ m, i.e., half the maximum separation present in the dataset.
Separate correlograms were drawn for samples taken during the same year, and samples separated in time by one year, in order to assess the degree to which the soil community changes over the course of one year.

Additionally, empirical distance-decay curves were generated by plotting mean community dissimilarity as a function of spatial distance, and fit to an exponential model of the form given by @legendre2012 using the `nls` function in R.
Points in the empirical distance-decay curve were weighted by the number of comparisons within the distance class and the inverse of the distance for the purposes of model fitting.
For datasets where the Mantel correlogram indicated spatial correlation between samples taken in separate years, the model was re-fit with an additional term to represent temporal correlation:
$$D = C_0 + C_1\left[1 - \exp\left(-3  \left(\frac{d}{a_d} + \frac{t}{a_t}\right)\right)\right]$$
where $D$, $d$, and $t$ represent the community dissimilarity, spatial distance, and time lag between samples, respectively, and the parameters are $C_0$, the community dissimilarity from replicate samples ("nugget"); $C_0 + C_1$, the community dissimilarity at long distances ("sill"); $a_d$ the spatial range at which the community dissimilarity has moved 95% of the way from "nugget" to "sill"; and $a_t$, the equivalent temporal range.
The 95% confidence intervals were calculated for the spatial and temporal range parameters by profiling using the `MASS` package in R.

# Results

<!-- R2: The Results section has many instances of the data "...are shown in Fig/Table..." This is difficult to read and I strongly suggest you summarize the findings in the text. In some cases it simply means deleting these sentences and adding the Fig/Table reference after the following sentence. It would be more helpful to talk about the overall trends here rather than listing all of the % values. -->

DNA concentrations after extraction and PCR, as well as sequencing reads for PacBio and Illumina, are shown per sample in Figure \@ref(fig:sample-depth).
Samples from Ang in 2015 yielded low quantities of DNA, poor PCR performance, and ultimately very few sequencing reads, especially in the long amplicon library, where only one sample produced more than 100\ reads.
Consequently, Ang samples were excluded from spatial analysis, although they were retained for denoising, phylogenetic reconstruction, and taxonomic assignment.

The number of sequencing reads and ASVs at each stage in the bioinformatics pipeline are shown in Table\ \@ref(tab:bioinfo).
Sequencing with PacBio RS II yielded more than twice as many raw reads for long amplicons as for short amplicons, with approximately 125 thousand and 50 thousand reads, respectively.
Ion Torrent Ion S5 and Illumina MiSeq yielded substantially more reads, with 20.7 million and 10.8 million, respectively.
<!-- R3: 20.7 M and 10.8M reads for only 25 samples is a lot of read depth! In the future you can go way lower than that to save money. I would multiplex in hundreds of samples for 20.8 M reads.-->
Demultiplexing, primer trimming, and quality filtering reduced these totals by `r (1-readd(filtercounts_full)/readd(rawcounts))["pb_500"] %>% percent(format = "d")` for PacBio long amplicons, but only by `r (1-readd(filtercounts_full)/readd(rawcounts))["pb_483"] %>% percent(format = "d")` for PacBio short amplicons, resulting in a similar number of filtered reads for the two strategies.
Losses in demultiplexing, trimming, and quality filtering were intermediate for Ion Torrent and Illumina, with `r (1-readd(filtercounts_full)/readd(rawcounts))[c("is_057", "SH-2257")] %>% percent(format = "d") %>% text_list()` loss, respectively.
In contrast, extraction of only the ITS2 region before quality filtering resulted in the loss of `r percent(1-readd(filtercounts_ITS2)['pb_500']/readd(rawcounts)['pb_500'], format = "d")` of trimmed long amplicon PacBio reads, `r percent(1-readd(filtercounts_ITS2)['pb_483']/readd(rawcounts)['pb_483'], format = "d")` of trimmed short amplicon PacBio reads, and `r percent(1-readd(filtercounts_ITS2)['is_057']/readd(rawcounts)['is_057'], format = "d")` of trimmed Ion Torrent reads.
This represented greater loss of PacBio short amplicon reads, but less loss of PacBio long amplicon reads and Ion Torrent reads.

Almost all of the short amplicons from all three technologies were between 240 and 375\ bp long (Figure\ \@ref(fig:full-length)a).
Although the length profile of the three sequencing runs were similar, Illumina MiSeq had the largest fraction of reads near the top of the range, followed by Ion Torrent Ion S5 and PacBio RS II (Figure\ \@ref(fig:full-length)b).
The difference in length distributions was statistically significant due to the large sample size (Kruskal-Wallis statistic = `r readd(length_kw)$statistic %>% formatC(format = "e", digits = 2)`, $p < 2.2\times10^{-16}$),
<!-- S3: 2 decimal points is sufficient-->
but the difference between means was fairly small, with mean amplicon lengths of `r filter(readd(reads_table), region == "short") %>% group_by(seq_run) %>% summarize(length = weighted.mean(length, reads)) %>% deframe() %>% magrittr::extract(c("pb_483", "is_057", "SH-2257")) %>% text_list(digits = 0)` bp for PacBio, Ion Torrent, and Illumina, respectively.
In contrast, the length of the long amplicon reads varied widely, from `r filter(readd(reads_table), region == "long") %$% range(length) %>% paste(collapse = " to ")` bp, with a mean of `r filter(readd(reads_table), region == "long") %$% weighted.mean(length, reads) %>% round()` bp.

The length distribution of the different regions extracted from the long amplicon are shown in Figure\ \@ref(fig:region-lengths).
ITS1 showed the greatest length variability (mean $\pm$ standard deviation: `r readd(reads_table) %>% filter(region == "ITS1", seq_run == "pb_500") %$% reldist::wtd.mean(length, weight = reads) %>% round()` $\pm$ `r readd(reads_table) %>% filter(region == "ITS1", seq_run == "pb_500") %$% reldist::wtd.var(length, weight = reads) %>% sqrt() %>% round()`\ bp), followed by ITS2 (`r readd(reads_table) %>% filter(region == "ITS2", seq_run == "pb_500") %$% reldist::wtd.mean(length, weight = reads) %>% round()` $\pm$ `r readd(reads_table) %>% filter(region == "ITS2", seq_run == "pb_500") %$% reldist::wtd.var(length, weight = reads) %>% sqrt() %>% round()`\ bp) and the variable regions in LSU (D2: `r readd(reads_table) %>% filter(region == "D2", seq_run == "pb_500") %$% reldist::wtd.mean(length, weight = reads) %>% round()` $\pm$ `r readd(reads_table) %>% filter(region == "D2", seq_run == "pb_500") %$% reldist::wtd.var(length, weight = reads) %>% sqrt() %>% round()`\ bp; D3: `r readd(reads_table) %>% filter(region == "D3", seq_run == "pb_500") %$% reldist::wtd.mean(length, weight = reads) %>% round()` $\pm$ `r readd(reads_table) %>% filter(region == "D3", seq_run == "pb_500") %$% reldist::wtd.var(length, weight = reads) %>% sqrt() %>% round()`\ bp; D1: `r readd(reads_table) %>% filter(region == "D1", seq_run == "pb_500") %$% reldist::wtd.mean(length, weight = reads) %>% round()` $\pm$ `r readd(reads_table) %>% filter(region == "D1", seq_run == "pb_500") %$% reldist::wtd.var(length, weight = reads) %>% sqrt() %>% round()`\ bp).
Approximately `r (readd(reads_table) %>% filter(region == "LSU4", seq_run == "pb_500", length > 200) %$% sum(reads)/readd(reads_table) %>% filter(region == "LSU4", seq_run == "pb_500") %$% sum(reads)) %>% signif(1) %>% percent()` of reads included an intron of 40--60 bp in the LSU4 region, not visible in Figure\ \@ref(fig:region-lengths) due to rarity.
Except for these sequences, all conserved regions of LSU, as well as 5.8S, displayed very little size variation, as expected, with standard deviations < 2 bp.

*Agaricus bisporus*, the positive control, was represented by a single ASV in the positive control samples for both long- and short-amplicon PacBio datasets, and in the Ion Torrent dataset.
*A. bisporus* was represented by two ASVs in the Illumina dataset, which differed at one base pair (`r Biostrings::readDNAStringSet(here::here("data/clusters/ITS2.fasta.gz")) %>% magrittr::extract(startsWith(names(.), "f2ab49e8")) %>% Biostrings::width() %>% unique() %>% divide_by(1, .) %>% subtract(1, .) %>% percent(digits = 1, format = "f")` similarity in ITS2).
The abundance of the second ASV was `r readd(agaricus_reads) %>% as.data.frame() %>% filter(rowSums(.) > 0) %>% {.[["f2ab49e8"]]/.[["39d80c19"]]} %>% keep(~. > 0) %>% percent(format = "f", digits = 1) %>% paste(collapse = " and ")` that of the primary *A. bisporus* ASV in the two Illumina positive controls.
The consistency of this ratio across replicate positive controls suggests that it represents true inter-copy variation within the specimen, rather than sequencing or PCR error.
Despite higher total sequencing depth, this ASV was not identified from the Ion Torrent dataset.

*A. bisporus* sequences represented `r readd(pos_control_data) %>% filter(complete.cases(.)) %>% group_by(seq_run) %>% summarize(read_frac = sum(pc_reads) / sum(all_reads)) %>% deframe() %>% magrittr::extract(c("pb_500", "pb_483", "SH-2257", "is_057")) %>% percent(format = "f", digits = 2) %>% text_list()` of non-control reads, in the PacBio long, PacBio short, Illumina, and Ion Torrent datasets, respectively, giving similar estimates for the rate of tag-switching for all technologies.
These reads were excluded from community analyses.
<!-- S3: so to clarify, everything was PCR amplified with a primer that had barcodes and adaptor attached to either forward or reverse primer? I was having trouble following this during you PCR description section.-->

## Reproducibility of sequence capture using different technologies

(ref:venn-s) Venn diagrams of shared ASVs and OTUs between different sequencing technologies

(ref:venn) Venn diagrams showing shared ITS2-based ASVs (*a*, *c*) and 97% OTUs (*b*, *d*) between different sequencing technologies from the same short amplicon library (*a*, *b*), and between long and short amplicon libraries (*c*, *d*).
In each region, the number of ASVs/OTUs is given above, while the fractions of reads for each sequencing strategy are shown below.
For short amplicons in *c* and *d*, ASV/OTU counts reflect detection by any of the three technologies, and read counts represent the mean fraction of reads across the three technologies.
<!-- S2: The values for Ion Torrent are 1-3 orders of magnitude higher than other methods. This to me is an important finding that is not mentioned(!) -->
<!-- s3: I recommend using “VennDiagram” package in R. It allows you to make Venn diagrams with circle sizes that change based on how many samples it represents. That is way easier to interpret than one with circles that are the same size regardless. I also don’t really understand what the decimals represent – what do you mean by the fraction of reads for each sequencing technology? For example in figure C, is 601 sequences 79% of the long amplicon reads? Please clarify. -->

```{r venn, fig.cap="(ref:venn)", fig.scap = "(ref:venn-s)", fig.show="hold", out.width = "70%", fig.height = 4, fig.align="center", results="asis"}
venn_A <- venndata(readd(asv_table), ASVs, cols = c("pb_483", "SH-2257", "is_057")) %>%
  vennplot_data(ASVs) %>% {
    plot(
      select(., set, ASVs) %>% deframe() %>% replace_na(0) %>% eulerr::venn(),
      legend = FALSE,
      labels = FALSE,
      quantities = .$label,
      main = "a",
      fills = c("grey85", "lightblue", "lightcoral"))
  }

venn_B <- venndata(readd(otu_table), OTUs, cols = c("pb_483", "SH-2257", "is_057")) %>%
  vennplot_data(OTUs) %>% {
    plot(
      select(., set, OTUs) %>% deframe() %>% replace_na(0) %>% eulerr::venn(),
      legend = FALSE,
      labels = FALSE,
      quantities = .$label,
      main = "b",
      fills = c("grey85", "lightblue", "lightcoral")
    )
  }
ggpubr::ggarrange(venn_A, venn_B, ncol = 2)

grid::grid.legend(
  labels =  c("PacBio Short", "Illumina", "Ion Torrent"),
  nrow = 1,
  pch = 21,
  gp = grid::gpar(fill = c("grey85", "lightblue", "lightcoral")),
  vp = grid::viewport(y = unit(0.05, "npc"))
)

grid::grid.newpage()

venn_C <- readd(asv_table) %>%
  mutate_if(is.numeric, ~ . / sum(.)) %>%
  mutate(long = pb_500,
         short = rowMeans(.[,c("pb_483", "SH-2257", "is_057")])) %>%
  select(seq, long, short) %>%
  venndata(ASVs, cols = c("long", "short")) %>%
  vennplot_data(ASVs) %>% {
    plot(
      select(., set, ASVs) %>% deframe() %>% replace_na(0) %>% eulerr::venn(),
      legend = FALSE,
      labels = FALSE,
      quantities = .$label,
      main = "c",
      fills = c("white", "slateblue1"))
  }

venn_D <- readd(otu_table) %>%
  mutate_if(is.numeric, ~ . / sum(.)) %>%
  mutate(long = pb_500,
         short = rowMeans(.[,c("pb_483", "SH-2257", "is_057")])) %>%
  select(seq, long, short) %>%
  venndata(OTUs, cols = c("long", "short")) %>%
  vennplot_data(OTUs) %>% {
    plot(
      select(., set, OTUs) %>% deframe() %>% replace_na(0) %>% eulerr::venn(),
      legend = FALSE,
      labels = FALSE,
      quantities = .$label,
      main = "d",
      fills = c("white", "slateblue1")
    )
  }
ggpubr::ggarrange(venn_C, venn_D, ncol = 2)
grid::grid.legend(
  labels =  c("Long amplicon", "Short amplicon"),
  nrow = 1,
  pch = 21,
  gp = grid::gpar(fill = c("white", "slateblue1")),
  vp = grid::viewport(y = unit(0.15, "npc"))
)
```

(ref:read-compare-s) Comparison between read numbers for different sequencing strategies

(ref:read-compare) by ASV (**a**) and 97% OTU (**b**). ASVs/OTUs which were detected by one sequencing strategy but not the other are plotted as tick marks along the axes. Dashed line represents a constant ratio of read numbers. The blue line is a LOESS smooth of the data, with associated uncertainty in grey shading. $R^2$ value displayed is for log-transformed non-zero read numbers.
<!-- R2: Please provide axis labels e.g. panel (a) y-axis is "Count of ASVs" (?) -->
<!-- R3: I’m not sure why this is useful, and I don’t think it warrants so much space in the document. It can be moved to the supplements. To me, a correlation of species richness by read technology might be interesting, or if you had some sort of treatment effect you were looking for and all of the methods gave you the same inferences of treatment effects, then that would be more interesting. I would prefer to see this replaced by correlation of alpha diversity.

I also think species accumulation curves between sequencing technologies would be interesting to see. -->

```{r read-compare, fig.cap="(ref:read-compare-s), (ref:read-compare)", fig.scap="(ref:read-compare-s)", fig.height=8}
ggpubr::ggarrange(
  read_comparison(
    readd(multi_table),
    readd(comparisons),
    type = "ASV"
  ),
  read_comparison(
    readd(multi_table),
    readd(comparisons),
    type = "OTU"
  ),
  ncol = 1,
  labels = "auto",
  label.x = 0.9
)
```

The majority of abundant ASVs and OTUs were captured by all sequencing strategies used (Figure\ \@ref(fig:venn)).
ASVs shared between all datasets represented `r paste(range(as.numeric(unlist(readd(venn_ASV)["1111", c("reads_frac_pb_500", "reads_frac_pb_483", "reads_frac_SH-2257", "reads_frac_is_057")]))*100), collapse = "--")`% of the reads for the long and short PacBio datasets, Illumina dataset, and Ion Torrent dataset, respectively.
These fractions increased to `r paste(range(as.numeric(unlist(readd(venn_OTU)["1111", c("reads_frac_pb_500", "reads_frac_pb_483", "reads_frac_SH-2257", "reads_frac_is_057")]))*100), collapse = "--")`% when differences at the intra-species scale were removed by clustering the ASVs into 97% OTUs.
In particular, `r text_list(percent(colSums(mutate_all(readd(venn_OTU)[c("0111", "1111"), c("reads_frac_pb_483", "reads_frac_SH-2257", "reads_frac_is_057")], as.numeric), na.rm = TRUE)))` of reads in the PacBio, Illumina, and Ion Torrent short-amplicon datasets belonged to OTUs shared between all three datasets.
In contrast, `r percent(readd(venn_ASV)["1000", "reads_frac_pb_500"])` of reads in the long PacBio dataset belonged to ASVs which were unique to that dataset, and the fraction only reduced to `r percent(readd(venn_OTU)["1000", "reads_frac_pb_500"])` after OTU clustering.
<!-- R1: Is this due to the much greater length causing a different ASV than a shorter ASV, even if they have an identical sequence over the same region? That seems to be the case, as the clustering did not really lower the difference between the unique long and short read ASVs at all (going from 21% to 20%).  More depth of explanation in the discussion of this issue, maybe with a couple of examples of what is driving this, would be helpful.-->
Complete tabulations of the number of ASVs and OTUs shared between the different sequencing strategies are shown in Supplementary Tables \@ref(tab:venn-table-asv) and \@ref(tab:venn-table-otu), respectively.

Figure\ \@ref(fig:read-compare) shows the correspondence between the read count for different ASVs (\@ref(fig:read-compare)a) and OTUs(\@ref(fig:read-compare)b) in the different technologies, where shared ASVs/OTUs are plotted as circles, and unshared OTUs are plotted as lines along the axes.
In all cases, the read counts for shared ASVs and OTUs were correlated, with a minimum $R^2$ value of `r compR2("pb_500", "is_057", "ASV")`.
Correlations between read counts for the three technologies using the short amplicon library were increased by OTU clustering
(`r compR2("pb_483", "SH-2257", "ASV")` to `r compR2("pb_483", "SH-2257", "OTU")`,
`r compR2("pb_483", "is_057", "ASV")` to `r compR2("pb_483", "is_057", "OTU")`, and
`r compR2("SH-2257", "is_057", "ASV")` to `r compR2("SH-2257", "is_057", "OTU")`,
for PacBio vs. Illumina, PacBio vs. Ion Torrent, and Illumina vs. Ion Torrent, respectively), but not between the long amplicon library and short amplicon library
(`r compR2("pb_500", "pb_483", "ASV")` to `r compR2("pb_500", "pb_483", "OTU")`,
`r compR2("pb_500", "SH-2257", "ASV")` to `r compR2("pb_500", "SH-2257", "OTU")`, and
`r compR2("pb_500", "is_057", "ASV")` to `r compR2("pb_500", "is_057", "OTU")`, for PacBio long amplicon reads vs. PacBio, Illumina, and Ion Torrent short amplicon reads, respectively).

## Taxonomic assignment

For all sequencing datasets and taxonomic assignment protocols, a higher proportion of reads was assigned than of ASVs, indicating that common ASVs were more likely to be identified than rare ASVs (Figure\ \@ref(fig:tax-chart)).
<!-- R3: do you mean taxonomically identified? -->
A greater fraction of ITS reads and ASVs were assigned using the Unite database than the Warcup database across sequencing technologies, amplicons, algorithms, and taxonomic ranks.
At most taxonomic ranks, the RDPC algorithm assigned the greatest fraction of reads and ASVs, followed by SINTAX, and then IDTAXA.

(ref:tax-chart-s) Summary of taxonomic assignments

(ref:tax-chart) Fraction of ASVs (left) and reads (right) assigned to each taxonomic rank, for different sequencing technologies (PacBio RS II, Illumina MiSeq, Ion Torrent Ion S5), amplicons (Long, Short), reference databases (Unite, Warcup, RDP), and assignment algorithms (PHYLOTAX, Consensus, RDPC, SINTAX, IDTAXA).
Consensus and PHYLOTAX assignments are based on the consensus of RDPC, SINTAX, and IDTAXA, using all available databases and, in the case of PHYLOTAX, phylogenetic information.
These two methods are plotted in each column to compare with results for the individual databases.
<!-- R2: Overall patterns are the same thus I suggest presenting only reads or ASVs but not both. Y-axis labels are needed as in Fig. 2. It is interesting that it is almost alinear decline-->
<!-- R3: the barplots are not big enough to readily distinguish. Perhaps you can make the ASVs vs Reads vertical rather than horizonal and horizontally enlarge your image so it’s easier to read.-->

```{r tax-chart, fig.cap = "(ref:tax-chart)", fig.scap = "(ref:tax-chart-s)"}
readd(tax_chart_plot)
```

(ref:fungi-comm-s) Taxonomic composition of fungal community at the class level

(ref:fungi-comm) Values represent the fraction of all ASVs, OTUs, or reads which were assigned to kingdom Fungi.
Assignments based on PHYLOTAX.
Classes which represented less than 2% of reads, OTUs, and ASVs in all datasets are grouped together as "other".

```{r fungi-comm, fig.cap = "(ref:fungi-comm-s). (ref:fungi-comm)", fig.scap = "(ref:fungi-comm-s)", fig.width=7}
readd(taxon_reads) %>%
  filter(
    kingdom == "Fungi",
    !(Algorithm == "Consensus" & region == "ITS"),
    Algorithm == "PHYLOTAX"
    ) %>%
  group_by(tech, amplicon) %>%
  mutate(reads = reads/sum(reads), ASVs = 1/n()) %>%
  pivot_longer(cols = c("reads", "ASVs"), names_to = "type", values_to = "reads") %>%
  bind_rows(
    filter(., type == "ASVs", label == OTU) %>%
      mutate(label = OTU) %>%
      group_by(tech, amplicon) %>%
      mutate(reads = 1/n(), type = "OTUs")
  ) %>%
  taxon_plot(
    rank = class,
    datasets = datasets,
    cutoff = 0.02,
    y = reads,
    x = type,
    facets = vars(tech, amplicon)
  ) +
  ylab("Fraction of Community") +
  xlab(NULL)
```

Taxonomic composition of the sequenced soil fungal community at the class level is is summarized in Figure\ \@ref(fig:fungi-comm) and as a heat tree [@foster2017] in Figure\ \@ref(fig:taxa).
The ML tree for fungal ASVs, along with taxonomic assignments, is shown in Supplementary File 3.
According to the PHYLOTAX assignments, Fungi represented `r percent(readd(fungi_asv)["PacBio_Long"], format = "f", digits = 0)` of the ASVs and `r percent(readd(fungi_read)["PacBio_Long"], format = "f", digits = 0)` of the reads in the long amplicon library, compared to `r paste(percent(range(readd(fungi_asv)[c("PacBio_Short", "Illumina_Short", "Ion.Torrent_Short")]), format = "f", digits = 1), collapse = "–")` of the ASVs and `r paste(percent(range(readd(fungi_read)[c("PacBio_Short", "Illumina_Short", "Ion.Torrent_Short")]), format = "f", digits = 1), collapse = "–")` of the reads in the short amplicon library.
Measured fungal community composition at the class level varied significantly between amplicons (PERMANOVA with 9999 permutations, $p<`r formatC(readd(tech_class_adonis)["amplicon", "p.value"], format = "f")`$, $R^2=`r formatC(readd(tech_class_adonis)["amplicon", "R2"], format = "f", digits = 3)`$), but only marginally between sequencing technologies ($p=`r formatC(readd(tech_class_adonis)["tech", "p.value"], format = "f")`$, $R^2=`r formatC(readd(tech_class_adonis)["tech", "R2"], format = "f", digits = 3)`$).
The majority of variation was spatiotemporal (i.e., between samples; $p<`r formatC(readd(tech_class_adonis)["paste(site, x, year)", "p.value"], format = "f")`$, $R^2=`r formatC(readd(tech_class_adonis)["paste(site, x, year)", "R2"], format = "f", digits = 2)`$), but once this variation was removed, the remaining effect consisted of a clear bias against Sordariomycetes in the long amplicon dataset (Figures\ \@ref(fig:fungi-comm) and \@ref(fig:ppcoa)).
<!-- R1: Any thoughts on why this clade in particular was well represented? For example, do you think this a primer bias issue or is there some strange secondary structure in the Sordariomycetes rRNA operon that results in their removal during bioinformatic filtering? -->

Fungi categorized as ECM made up `r percent(readd(ecm_asv)["pb_500"], format = "f", digits = 1)` of ASVs and `r percent(readd(ecm_read)["pb_500"], format = "f", digits = 1)` of reads in the long amplicon library, and `r paste(percent(range(readd(ecm_asv)[c("pb_483", "SH-2257", "is_057")]), format = "f", digits = 1), collapse = "–")` of the ASVs and `r paste(percent(range(readd(ecm_read)[c("pb_483", "SH-2257", "is_057")]), format = "f", digits = 1), collapse = "–")` of the reads in the short amplicon library (Figure\ \@ref(fig:ecm)).
Although amplicon length had a significant effect on ECM community composition at the family level, the explained variation was very low (PERMANOVA with 9999 permutations, $p=`r formatC(readd(tech_ecm_fam_adonis)["amplicon", "p.value"], format = "f")`$, $R^2=`r formatC(readd(tech_ecm_fam_adonis)["amplicon", "R2"], format = "f", digits = 3)`$), and the majority of variation was again spatiotemporal ($p<`r formatC(readd(tech_ecm_fam_adonis)["paste(site, x, year)", "p.value"], format = "f")`$, $R^2=`r formatC(readd(tech_ecm_fam_adonis)["paste(site, x, year)", "R2"], format = "f", digits = 2)`$).
Variation between sequencing technologies was not significant ($p=`r formatC(readd(tech_ecm_fam_adonis)["tech", "p.value"], format = "f", digits = 2)`$, $R^2=`r formatC(readd(tech_ecm_fam_adonis)["tech", "R2"], format = "f", digits = 4)`$).

## Spatial analysis

Results of spatial analysis based on the Bray-Curtis dissimilarity were qualitatively similar between the two amplicon libraries and between PacBio and Illumina sequencing, with significant autocorrelation at $p < 0.05$ for ranges of up to 2--3\ m for the total fungal community, and 1--2\ m for the ECM fungal community (Figure\ \@ref(fig:correlog)).
In both cases, the greatest correlation magnitudes were found with Illumina, followed by long amplicon PacBio.
The least spatial structure was detected with PacBio short amplicon sequencing.

The Bray-Curtis metric showed significant ($p < 0.05$) positive correlation when resampling at the same locations one year later (i.e., spatial distance of 0\ m, time lag of 1\ year), for both the total fungal and ECM fungal communities in the long amplicon library.
For the short amplicon library, although the general profile of the correlograms was similar, correlation at 0\ m and 1\ year was not significant, but there was a negative correlation at time lag of 1\ year and a distance of 1\ m for both sequencing technologies.
<!-- R3: what does this mean?-->
This puzzling negative correlation was significant in all correlograms based on short amplicon sequencing irrespective of technology.

In contrast to the Bray-Curtis distance, the weighted UniFrac distance showed very little spatial structure, with only the total fungal community in the 1\ m
distance class showing a significant correlation at $p < 0.05$.
No temporal correlation was found for the weighted UniFrac distance.

The best fit spatial ranges based on distance-decay curves vary between the different datasets by a factor of about 3, but there is overlap of the 95% confidence intervals for all of the Bray-Curtis spatial ranged in both the total fungal and ECM fungal communities, across amplicon libraries and sequencing technologies (Figure\ \@ref(fig:spatio), Table\ \@ref(tab:variofit)).
<!-- R3: please re-write to better explain what you mean. Do you mean that communities generally experience full compositional turnover at 3m?-->
Although a distance-decay model was fit for the weighted UniFrac distance applied to the total fungal community, the result was very poorly constrained, and a range of 0\ m, indicating no spatial structure, was included in the 95% confidence interval.


(ref:spatio-s) Distance-decay plot for community dissimilarities and spatio-temporal distance

(ref:spatio) Circles represent community data from short (top two rows) and long (bottom two rows) amplicon libraries, sequenced by Illumina MiSeq (top row) or PacBio RS II (bottom three rows).
Community dissimilarities are calculated using the Bray-Curtis dissimilarity for all datasets (top three rows) and using the weighted UniFrac dissimilarity for the long amplicon library, for which a phylogenetic tree could be constructed (bottom row).
The left column represents the full fungal community, and the right column only sequences identified as ECM.
The color of each circle represents the time lag between samples being compared (0 or 1\ year), and the size represents the number of comparisons for that spatial distance and time lag.
Lines are the best-fit lines for an exponential decay to max model.
The model was only fit for datasets where the Mantel test indicated a significant relationship between community dissimilarity and spatial (for the 0\ year timelag) or spatiotemporal (for the 1\ year time lag) distance.

```{r spatio, fig.cap = "(ref:spatio-s). (ref:spatio)", fig.scap = "(ref:spatio-s)"}
readd(variog_empir) %>%
  filter(algorithm == "PHYLO") %>%
  ggplot(aes(dist, gamma, color = timelag, size = n)) +
  ggnomics::facet_nested(amplicon + tech + metric ~ guild, scales = "free", nest_line = TRUE, space = "free_y") +
  # geom_hline(data = variog_sills, aes(yintercept = sills), linetype = "dashed", color = "gray50") +
  # geom_vline(data = variog_data, aes(xintercept = range), linetype = "dashed", color = "gray50") +
  # geom_jitter(shape = 46, alpha = 0.5) +
  # geom_line(size = 1.5, alpha = 0.5) +
  geom_point(alpha = 0.5) +
  # geom_smooth(data = variog_points, aes(fill = timelag), color = FALSE, alpha = 0.5) +
  geom_line(data = filter(readd(variog_fit), timelag == "1", algorithm == "PHYLO"), size = 0.5, color = "darkcyan", alpha = 0.8) +
  geom_line(data = filter(readd(variog_fit), timelag == "0", algorithm == "PHYLO"), size = 0.5, color = "red1", alpha = 0.8) +
  scale_x_continuous(expand = expand_scale(0, 0), limits = c(0, 25), name = "Distance (m)") +
  scale_color_discrete(name = "Time lag") +
  scale_size_continuous(name = "Comparisons") +
  scale_y_continuous(name = "Community dissimilarity") +
  theme(strip.background = element_blank())
```

# Discussion

<!-- R2: The Discussion section is mostly descriptive, largely repeating the Results (e.g., lines 569-574; 578-581;589-593; 596-598; 600-606) with very little real discussion. -->

## Reconstruction of long amplicons from denoised subregions

ASV recovery for long amplicons using DADA2 was dramatically improved (`r percent(readd(region_table)$reads[readd(region_table)$region == "long"]/as.integer(gsub(" ", "", readd(demuxcounts)["pb_500"])), format = "d")` to `r percent((readd(asvcounts) %>% filter(seq_run == "pb_500", step == "long") %>% pull(reads))/as.integer(gsub(" ", "", readd(demuxcounts)["pb_500"])), format = "d")` of reads) by denoising homologous subregions independently using the `LSUx` and `tzara` packages.
<!-- Why does the difference range from 12-76% This is a wide range. I would like to know what the authors think is the reason for this. Under what circumstances might we expect 12% and what 76% and why? What is the contribution beyond that of other studies that separated long reads into constituent genes regions and examined these separately? -->
Although newer sequencing platforms from PacBio (Sequel and Sequel II) feature increased sequencing depth and lower error rate compared to the RS II, long sequences inherently require much more sampling depth to identify ASVs.
Thus, `tzara` should increase ASV recovery from these platforms as well.
It may also be adaptable to Oxford Nanopore sequencing, which has hitherto posed difficulties for application to complex community metabarcoding [@loit2019].

## Comparison of sequencing strategies

The three sequencing technologies gave similar results for the short amplicon library, the major difference being in sequencing depth.
Although a greater fraction of PacBio raw reads were ultimately mapped to ASVs (`r percent(readd(bioinf_table)["ITS2","PacBio_Short_reads"]/readd(rawcounts)["pb_483"], format = "f", digits = 0)`) compared to Illumina (`r percent(readd(bioinf_table)["ITS2","Illumina_Short_reads"]/readd(rawcounts)["SH-2257"], format = "f", digits = 0)`) or Ion Torrent (`r percent(readd(bioinf_table)["ITS2","Ion Torrent_Short_reads"]/readd(rawcounts)["is_057"], format = "f", digits = 0)`),
the latter two technologies provided much greater sequencing depth for a similar cost, allowing a greater diversity of rare ASVs to be recovered.
<!--R3: this could be visualized with a species accumulation curve-->

DADA2 denoising may perform differently on different technologies (or perhaps sequencing runs), indicated by the fact that clustering ASVs at 97% led to substantially higher correspondence between both the set of sequences recovered from the same library by different technologies (Figure\ \@ref(fig:venn)) and the read counts for each sequence (Figure\ \@ref(fig:read-compare)).
The large number of ASVs unique to Ion Torrent, while only the Illumina dataset recovered an apparent intragenomic variant in the positive control sample, suggests that DADA2 may not control sequencing error as effectively in Ion Torrent sequences as in Illumina, for which it was developed [@callahan2016].
<!-- R2: I understand that dada2 may not work as well, but why not?-->

Although the longer read length capabilities of PacBio would allow recovery of longer ITS2 sequences than the other two technologies, PacBio did not recover any ITS2 fragments longer than those recovered by Illumina and Ion Torrent.
Notably, neither long nor short amplicon sequencing recovered any sequences identifiable to *Cantharellus*, an ECM genus which is commonly observed at the study sites as fruitbodies (personal observation), but which is also known to have accelerated evolution in the rDNA [@moncalvo2006] and longer ITS regions than other fungi [@Feibelman_1994], making it an especially difficult target for metabarcoding.
Contrary to expectations, Illumina showed a slightly higher fraction of longer ITS2 sequences than Ion Torrent, which in turn showed slightly longer sequences than PacBio (Figures\ \@ref(fig:full-length) and \@ref(fig:ITS2-length)).

Of long amplicon reads, `r percent(readd(venn_ASV)["1000", "reads_frac_pb_500"])` belonged to ASVs which occurred only in the long amplicon dataset, and clustering at 97% similarity only reduced this fraction to `r percent(readd(venn_OTU)["1000", "reads_frac_pb_500"])`.
Additionally, ITS2 sequences extracted from the long amplicon dataset included some sequences that were much shorter than those recovered from the short amplicon datasets (Figure \@ref(fig:ITS2-length)).
Taxonomic assignments revealed that the majority of these non-shared sequences fall outside kingdom Fungi (Figure\ \@ref(fig:heattree-length-compare)), and that in particular the short ITS2 sequences are mostly Alveolates (Figure\ \@ref(fig:short-its)).
Within Fungi, the short amplicon datasets recovered more Sordariomycetes (Figures\ \@ref(fig:fungi-comm), \@ref(fig:ppcoa), and \@ref(fig:heattree-length-compare)).
Additionally, several smaller groups showed increased detection in either the long or short datasets, such as Tulasnellaceae and Pyronemataceae in the long amplicon dataset, and *Myerozyma* in the short amplicon datasets (Figures\ \@ref(fig:heattree-length-compare) and \@ref(fig:ecm-heattree)).
<!-- R3: what do you mean by smaller groups? Having fewer genera per class?-->
These differences may be due to primer mismatches in these taxa.
<!-- R1: Can this be quantified? For example, do we know how well for example LR5 selects for fungi relative to alveolates based on priming site matches? Additionally, were the short reads more common on one side of the “long read” than the other? For example, if all the short reads are common on the LR5 end of the long read, that might suggest using a different LSU primer would minimize the inclusion of alveolates and other non-target taxa.-->
<!-- R2: Primer mismatches are mentioned here, but only in relation to a few groups and as a final point in this section. But these are established primers whose biases are probably known and published. Do you results make sense? I would also need to see some discussion on how this bias might affect the study as a whole, because different platforms/reads used different primers. -->

## Taxonomic identification

<!-- R2: This section is weak and could be removed. -->
The RDP fungal training set and Unite performed comparably at taxonomic placement of long amplicon sequences.
The Warcup database placed notably fewer sequences at all taxonomic levels for all datasets, probably in part due to the fact that only fungal sequences are included.
However, even with this considered, IDTAXA performed very poorly with the Warcup database, placing <25% of ASVs to kingdom in all datasets.
<!-- R2: why is this the case?-->
IDTAXA placed fewer sequences than RDPC or SINTAX even with the other databases, but this is expected given its more conservative assignment of confidence scores [@murali2018a].
<!-- "This is expected" is not a very interesting conclusion in this case. I do not see how any other results could occur. Less conflict at lower taxonomic levels would only be a reflection of different taxonomies among the databases, and would not be anything biological. -->
  
@gdanetz2017 showed that a majority-rule consensus of three assignment algorithms can improve the fraction of sequences assigned as well as decrease the false assignment rate.
<!-- R1: See my earlier note about Palmer et al. (2018) showing a similar result. -->
Strict consensus rejects assignments whenever there is conflict between methods and should therefore provide more conservative taxonomic assignments than majority-rule consensus.
Here, we found that strict consensus also usually increases the number of assigned sequences relative to any single method, except at family and genus level identifications.
This suggests that different assignment algorithms and databases bring mostly complementary, non-contradictory information at higher taxonomic levels.
However, contradictory assignments between different methods is more common at lower taxonomic levels, which can be problematic because accurate assignment at the family or genus level is generally required for ecological guild assignment using FUNGuild.

For ASVs where a long amplicon sequence is available, PHYLOTAX uses phylogenetic relationships to resolve these disagreements in a principled manner.
For instance, `r filter(readd(tax_chart), tech == "Illumina", Algorithm == "Consensus", rank %in% c("genus", "family"), type == "Reads", reference == "Unite") %>% pull(frac) %>% percent(format = "f", digits = 0) %>% rev() %>% text_list()` of Illumina reads were assigned to genus and family, respectively, by the strict consensus of methods, but PHYLOTAX increased this fraction to `r filter(readd(tax_chart), tech == "Illumina", Algorithm == "PHYLOTAX", rank %in% c("genus", "family"), type == "Reads", reference == "Unite") %>% pull(frac) %>% percent(format = "f", digits = 0) %>% rev() %>% text_list()`.
This led to a corresponding increase in the fraction of reads assigned to a functional guild (Figure\ \@ref(fig:ecm)).

## Turnover rate

<!-- R2: This section is too descriptive, and would be greatly improved if reasons were given, even multiple possible reasons that lead to hypotheses. It appear that the only conclusion offered (UNIFRAC would be more suited to larger scales) could have been made without these data. -->
<!-- R3: the Unifrac results should be reported after the Bray-Curtis results, which are more interesting. They could also be condensed.-->

Weighted UniFrac did not reliably detect spatial structure within this relatively ecologically homogeneous community.
Although the Mantel test did show a small but significant positive autocorrelation in the fungal community at the smallest size category (1\ m; Figure\ \@ref(fig:correlog)), the distance-decay plot in Figure\ \@ref(fig:spatio) does not show any clear relationship.
The functional fit showed poor convergence, with a 95% confidence interval for spatial range of 0--5700\ m, indicating little evidence of spatial structure.
This is probably due to the majority of weighted branch length in the community being between the Pezizomycotina and Agaricomycetes (Figure\ \@ref(fig:taxa)), which are both well represented in the majority of samples.
UniFrac would be more suited at larger spatial scales and/or larger ecological gradients.
<!-- R1: A second way to try to understand this result is to see if the results differ based on weighting versus not weighting. Similarly, it would be nice to see is the distance-decay patterns are robust to abundance (read counts) versus incidence (pres/abs) based analyses. There are some in the fungal metabarcoding world who believe all high throughput sequencing (HTS) data should only be analyzed using presence absence data due to ITS copy number differences and other known issues (primer biases, etc). I don’t agree and I think the best way to push back on that critique is to show that the results of abundance- and incidence-based data classifications often yield similar patterns. The results of these suggested additional analyses could easily be incorporated as extra panels in Figure 5. -->

Mantel correlograms based on the Bray-Curtis dissimilarity (Figure\ \@ref(fig:correlog)) revealed spatial autocorrelation in the soil fungal community at distance classes $\le 3$ m for both Illumina and PacBio using long and short amplicons, and in the ECM fungal community at distance classes $\le 2$ m for Illumina and PacBio long amplicons, and $\le 1$ m for the PacBio short amplicons.
These results are similar to autocorrelation ranges found in previous work based on ECM root tips in temperate forests [@lilleskov2004; @pickles2012a].
@lilleskov2004 found autocorrelation only at ranges <2.6\ m at most sites using Sanger sequencing.
Similarly, @pickles2012a found autocorrelation at distances <3.4\ m based on T-RFLP analysis.
However, previous work in Miombo woodland, a similar ecosystem to the Soudanian woodland in this study, found autocorrelation at ranges <10\ m using Sanger sequencing of ECM root tips [@tedersoo2011], which was their smallest distance class.
<!-- R3: I would not start this line with however, because it agrees with the sentences before if <10 m is the smallest distance class-->

Distance-decay plots (Figure\ \@ref(fig:spatio), Table\ \@ref(tab:variofit)) gave substantially longer autocorrelation distances.
There was little variation in the results between the Illumina and long-amplicon PacBio datasets for both the total fungal community and the ECM community, with best fit estimates ranging from `r readd(variofit_table) %>% filter((amplicon == "Long") | (tech == "Illumina"), metric == "Bray-Curtis", term == "range") %$% range(estimate) %>% format(digits = 0) %>% paste(collapse = "--")`\ m.
The 95% confidence interval was substantially wider than this variation, generally covering a range of `r readd(variofit_table) %>% filter((amplicon == "Long") | (tech == "Illumina"), metric == "Bray-Curtis", term == "range") %$% range(c(conf.low, conf.high), na.rm = TRUE) %>% format(digits = 0) %>% paste(collapse = "--")`\ m.
All of these values are smaller than the 65\ m reported by @Bahram2013, also based on distance-decay curves from an ECM woodland habitat in Benin.
<!-- R1: Is the habitat presented in Bahram et al. (2013) in Benin differ in some fundamentally important way from the other in this study (less disturbance, higher ECM tree basal area?It would be nice to try to better understand these differences, even if it requires a bit of speculation by the authors.-->

The PacBio short amplicon dataset shows a longer spatial range, of 
`r readd(variofit_table) %>% filter(amplicon == "Short", tech == "PacBio", metric == "Bray-Curtis", term == "range", guild == "All Fungi") %$% range(estimate) %>% format(digits = 0) %>% unique() %>% paste(collapse = "--")`\ m for the total fungal community and
`r readd(variofit_table) %>% filter(amplicon == "Short", tech == "PacBio", metric == "Bray-Curtis", term == "range", guild == "ECM") %$% range(estimate) %>% format(digits = 0) %>% unique() %>% paste(collapse = "--")`\ m for the ECM community,
<!-- R1: 38-38m – is this a typo? FIXED-->
<!-- R3: is 38-38m a typo?-->
in both cases with wide confidence intervals spanning `r readd(variofit_table) %>% filter(amplicon == "Short", tech == "PacBio", metric == "Bray-Curtis", term == "range") %$% range(c(conf.high, conf.low), na.rm = TRUE) %>% unique() %>% format(digits = 0) %>% paste(collapse = "--")`\ m.
It is possible that the weaker fit for this dataset, which also showed weaker autocorrelation in the Mantel correlogram, is due to low sequencing depth.
<!-- R1: Could this explanation be tested more formally by rarefying the non-ECM dataset down to the level of the ECM dataset and seeing if the confidence intervals and range estimates change? It would be nice to separate likely methodological artifact from potential differences in biology (as ECM communities likely have many more basids, so less asexual propagule production). -->

The Bray-Curtis Mantel correlogram for both the total fungal and ECM communities from the long amplicon dataset show a significant positive correlation at 0\ m and 1\ year.
<!-- R3: This would be much more engaging to read, here and in other places, if you reword to report findings rather than stats “There was strong year to year correlation for both the total and ECM fungal communities (cite figure or stats)” -->
The spatiotemporal distance-decay fit estimated the temporal turnover range as `r readd(variofit_table) %>% filter(amplicon == "Long", tech == "PacBio", metric == "Bray-Curtis", term == "timerange", guild == "All Fungi") %$% range(estimate, na.rm = TRUE) %>% unique() %>% formatC(format = "f", digits = 1) %>% paste(collapse = "--")`\ years for the total fungal community and `r readd(variofit_table) %>% filter(amplicon == "Long", tech == "PacBio", metric == "Bray-Curtis", term == "timerange", guild == "ECM") %$% range(estimate, na.rm = TRUE) %>% unique() %>% formatC(format = "f", digits = 1) %>% paste(collapse = "--")`\ years for the ECM community, but with overlapping confidence intervals.
Both datasets from the short amplicon library showed a puzzling pattern with no autocorrelation at 0\ m and 1\ year, but a weak negative correlation at 1\ m and 1\ year.
The general shape of the correlograms were similar for long and short amplicon datasets.
We hypothesize that two different processes may be at work with differing spatiotemporal scales, whose superposition result in this pattern.
<!-- R2: This is one example of hypothesis formation, although these are very vague and could be more clearly presented - which scales and how would the different processes lead to this result? -->

## Conclusion

The choice of amplicon and sequencing technology did not seem to affect the results of the spatial analysis, provided sufficient sequencing depth.
However, the addition of long amplicon reads did allow the construction of a phylogenetic tree from the metabarcoding reads, which allowed refinement of taxonomic assignments.
<!-- R2: I don't see this as a valid conclusion, because this was not attempted with short reads. In principle it could be done. -->
DADA2 ASV yield was initially poor for long amplicons, but this was improved by developing a workflow for extraction of subregions, separate denoising, and then reconstruction of full-length unique sequences.
Together these approaches provide a hybrid approach using long-read sequencing to acquire long amplicon sequences for the local species pool, and cost-effective short-read sequencing to provide high sampling depth and sample number.

# Acknowledgements {-}

This project was funded by the Swedish research council FORMAS grant number 2014-01109.

Laboratory work including PCR and library pooling was performed by Dr. Ylva Strid.
<!-- R3: how come Dr Ylva Strid is not a co-author on the paper? Seems like she contributed substantially.-->

The authors would like to acknowledge support of the National Genomics Infrastructure (NGI) / Uppsala Genome Center and UPPMAX for providing assistance in massive parallel sequencing and computational infrastructure.
Work performed at NGI / Uppsala Genome Center has been funded by RFI / VR and and Science for Life Laboratory, Sweden.

Sequencing was performed by the SNP&SEQ Technology Platform in Uppsala.
The facility is part of the National Genomics Infrastructure (NGI) Sweden and Science for Life Laboratory.
The SNP&SEQ Platform is also supported by the Swedish Research Council and the Knut and Alice Wallenberg Foundation.

\singlespacing

\printbibliography

\doublespacing

# Data Accessibility {-}

 - Trimmed, demultiplexed sequencing reads have been deposited at the European Nucleotide Archive (ERA) under Project accession number PRJEB37385. Accession numbers are given in Supplementary Files 1 and 2.
 - Consensus ASV sequences will also be deposited at ENA prior to publication.
 - Nucleotide alignment and ML tree will be deposited at Dryad prior to final publication [@furneauxdata2020].
 - R packages `LSUx`, `tzara`, `phylotax`, and `FUNGuildR` are available on Github at https://github.com/brendanf/LSUx, https://github.com/brendanf/tzara, https://github.com/brendanf/phylotax, and https://github.com/brendanf/FUNGuildR.
These packages are currently being prepared for submission to CRAN/Bioconductor.
If they are not accepted prior to final publication, snapshots will be archived at Dryad.
 - FASTA-format files for the RDP, Warcup, and Unite reference databases with unified classifications, as well as scripts used to generate them, are available at https://github.com/brendanf/reannotate. The versions used in this paper will be archived at Dryad prior to publication.
 - Bioinformatics pipeline and analysis scripts are available at https://github.com/oueme-fungi/oueme-fungi-transect.

# Author Contributions {-}

Sampling was planned and carried out by BF, NSY, and MR.
Bioinformatics and data analysis were performed by BF with input from MB, AR, and MR.
Scripts and R packages were written by BF.
The manuscript was drafted by BF and MR.
All authors contributed to and approved the final version of the manuscript.

\processdelayedfloats