---
title: "A Tale of Two Transects"
subtitle: "Also Two Sequencing Technologies, and Two Amplicons"
author: 
  - Brendan Furneaux^1^,
    Roël Houdanon^2^,
    Mohammad Bahram^1^,
    Anna Rosling^3^,
    Nourou S. Yorou^2^,
    Martin Ryberg^1^
  - ^1^Program in Systematic Biology,
    Department of Organismal Biology,
    Uppsala University,
    Uppsala, Sweden
  - ^2^Research Unit in Tropical Mycology and Plant-Fungi Interactions,
    LEB,
    University of Parakou,
    Parakou, Benin
  - ^3^Program in Evolutionary Biology,
    Department of Ecology and Genetics,
    Uppsala University,
    Uppsala, Sweden
date: "`r format(Sys.time(), '%B %d, %Y')`"
bibliography:
  - "all.bib"
  - "R.bib"
csl: fungal-ecology.csl
panflute-filters: figurefilter
panflute-path: '../../python'
output:
  bookdown::pdf_document2:
    fig_crop: no
    keep_tex: yes
    latex_engine: lualatex
    includes:
      in_header: preamble.tex
  bookdown::word_document2: 
    keep_md: yes
---

```{r setup, include=FALSE}
library("knitr")
library("tikzDevice")
knitr::opts_chunk$set(dev="tikz",
                      # dev.args = list(pointsize=32),
                      echo=FALSE,
                      results='asis',
                      fig.show='asis',
                      external=TRUE)
options(tikzDefaultEngine = "luatex",
        tikzLualatexPackages = c(options("tikzXelatexPackages")$tikzXelatexPackages,
                                "\\usepackage[version=4]{mhchem}\n",
                                "\\usepackage{siunitx}\n",
                                "\\DeclareSIUnit\\year{a}\n"))

base.dir <- regmatches(getwd(), regexpr(".*oueme-fungi-transect", getwd()))
if (opts_chunk$get("dev") == "tikz") {
  # styles$Expression <- styles$LaTeX.Expression
  # styles$Name <- styles$LaTeX.Name
  # styles$Short.Name <- styles$LaTeX.Short.Name
  # styles$Unit <- paste0("\\si{", styles$LaTeX.Unit, "}")
}
write_bib(c("base", "vegan", "permute"), file = "R.bib", tweak = TRUE)
```
# Introduction

# Methods

## Sampling

Sampling was conducted at two sites approximately 30 km apart in the _Forêt Reservée de l'Ouémé Supérieur_ (Upper Ouémé Forest Reserve) in central Benin.
Both locations were located in woodlands dominated by the ECM host tree _Isoberlinia doka_ (Caesalpinioideae).
At each site, 25 soil samples were collected at intervals of 1 m.
For each sample, any coarse organic debris was removed from the soil surface and a sample of approximately 5cm×5cm×5cm was extracted with a knife blade.
Each sample was sealed in a plastic zipper bag and homogenized by shaking and manually breaking apart soil aggregations.
Approximately 50 mg total of soil was collected from two locations in the homogenized soil and placed in Xpedition lysis solution (Zymo Research) and lysed in the field using a handheld bead-beater (Terralyzer; Zymo Research).
The remainder of each sample was dried at 40°C with an electric dehydrator within 24-48 hours of collection for bulk chemical analysis.

  - Ouémé Forest reserve
  - Two locations in _Isoberlinia doka_ forest, approx. 30km apart
  - transect: 25m, sampled at interval of 1m
  - Sampled in May 2015 and June 2016
  - remove coarse debris (typically very little) and take ~5cm×5cm×5cm sample
  - homogenize by hand in ziplock bag
  - take $\approx$ 50mg, lysis in field with Terralyzer
  - remainder of sample dried at 40°C with electric dryer within 24-48 hours

## Soil Chemistry

  - Need to get methods from Nourou

## DNA extraction, PCR, and sequencing

  - lysis in field Zymo Research Xpedition kit
  - short amplicon : gITS7 -- ITS4 (multiplexing index on gITS7)
  - long amplicon : ITS1 -- LR5 (multiplexing index on both ends)
  - short amplicon : same library sequenced with PacBio RS II and Ion Proton S5
  - long amplicon : PacBio RS II
  
## Bioinformatics

  - Two workflows for short reads:
    - Mohammad's workflow in Pipecraft
    - Brendan's workflow using Snakemake/Drake
    
  - Mohammad's pipeline:
  
  - Brendan's pipeline:
      - Ion Torrent
          - used demultiplexing from SciLife (standard IonXpress)
          - primers removed with `cutadapt` [@martin2011]
      - PacBio
          - CCS basecalls re-done with `ccs` version ___ [@pacificbiosciences2019]
          - demultiplexed and primers removed with `cutadapt` [@martin2011]
      - Illumina MiSeq
          - pending
      - Region extraction with LSUx (available as R package on github and conda)
          - 5.8S located with Rfam [@kalvari2018] covariance model (CM) using `cmsearch` from Infernal 1.1.2 [@nawrocki2013]
          - read trimmed at 5' end of 5.8S; remaining read aligned to modified 5.8S + LSU (=32S) CM from Ribosomal Data Project (RDP) [@glockner2017, @cole2014] with annotations for different LSU domains (definitions from @michot1984 and @raue1988), using `cmalign` from Infernal 1.1.2
          - 5.8S, ITS2, D1, D2, D3, and conserved LSU domains (called LSU1, LSU2, LSU3, LSU4) extracted from alignment using annotations; ITS1 is everything before 5.8S.
      - extracted regions independently quality filtered by max expected error using `dada2::filterAndTrim()`
      - dereplicated using `dada2::derepFastq()`
      - denoised with `dada2::dada()`
      - de novo chimera removal using `dada2::removeBimeraDenovo()`
      - full length amplicons reconstructed from ASVs of extracted domains (r package tzara -- available on github and conda)
          - when a read is mapped to ASVs for all regions, they are concatenated.
          - bimera detection on each trio of variable-conserved-variable domains (e.g., ITS1-5.8S-ITS2, ITS2-LSU1-D1, etc.); reads which are detected at bimera for any trio are discarded.

      - taxonomy assignment
          - three methods: 
              - IDTAXA [@murali2018]
              - SINTAX [@edgar2016a] as implemented in VSEARCH v2.9.1 [@rognes2016]
              - RDP Naïve Bayesian Classifier (NBC) [@wang2007] as implemented in DADA2
          - three databases:
              - UNITE version 8 [@nilsson2019a]
                  - extracted ITS1 and ITS2 using ITSx [@Nilsson2010; @bengtsson-palme2013]
                  - trained three classifiers: full (for ITS, 5_8S, full amplicons), ITS1, ITS2.
              - RDP fungal LSU training set 11.5 [@cole2014]
                  - used to identify long amplicon, LSU
                  - taxonomy for non-fungi looked up from NCBI based on accession numbers
              - RDP Warcup fungal ITS training set [@wang2007]
                  - ITS1 and ITS2 extracted as for Unite;
          - taxonomic classification in the three datasets were standardized to @tedersoo2017c
              - full classification to genus level available in supplementary material
              - apparently used by Unite for nonfungi; @tedersoo2018 used by Unite for fungi
              - groups monophyletic
              - SINTAX and NBC require a ranked system, as opposed to more accepted classifications like @adl2019
      - ecological guild assigned using FUNGuild [@nguyen2016funguild]
      - alignment as RNA in DECIPHER (iterative progressive alignment with RNS secondary structure prediction) [@wright2015]
      - tree using RAxML 8.2 [@stamatakis2014]; GTR+GAMMA; rapid bootstrapping with MRE-IGN stopping criterion [@pattengale2010].

<!--
          - preliminary alignment of 32S to RDP covariance model with `cmalign` from Infernal (aligns only conserved regions)
          - concatenate ITS1 to beginning of alignment
          - create guide tree using alignment of only conserved, gap-free columns in preliminary alignment, using UPGMA in DECIPHER [@wright2016]
          - realign nonconserved sites using MLocARNA (progressive simultaneous alignment and folding) from LocARNA 2.0.0RC8 [@will2007; @will2012; @will2013]. -->

      - alignment/tree for long amplicons built iteratively with PASTA[@mirarab2014; @mirarab2014a]. Fastree tree, mafft align, OPAL merge, RaxML final tree
          
## Statistics

# Results

# Discussion

# Bibliography {-}

<div id="refs"></div>