% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
%
\documentclass[
]{article}
\usepackage{lmodern}
\usepackage{amssymb,amsmath}
\usepackage{ifxetex,ifluatex}
\ifnum 0\ifxetex 1\fi\ifluatex 1\fi=0 % if pdftex
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math}
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\IfFileExists{bookmark.sty}{\usepackage{bookmark}}{\usepackage{hyperref}}
\hypersetup{
  pdfauthor={Brendan Furneaux1,; Mohammad Bahram2,3,; Anna Rosling4,; Nourou S. Yorou5,; Martin Ryberg1; 1Program in Systematic Biology, Department of Organismal Biology, Uppsala University,; Uppsala, Sweden; 2Department of Ecology, Swedish University of Agricultural Sciences, Uppsala, Sweden; 3Institute of Ecology and Earth Sciences, University of Tartu, Tartu, Estonia; 4Program in Evolutionary Biology, Department of Ecology and Genetics, Uppsala University,; Uppsala, Sweden; 5Research Unit in Tropical Mycology and Plant-Fungi Interactions, LEB,; University of Parakou, Parakou, Benin},
  hidelinks,
  pdfcreator={LaTeX via pandoc}}
\urlstyle{same} % disable monospaced font for URLs
\usepackage[margin=1in]{geometry}
\usepackage{longtable,booktabs}
% Correct order of tables after \paragraph or \subparagraph
\usepackage{etoolbox}
\makeatletter
\patchcmd\longtable{\par}{\if@noskipsec\mbox{}\fi\par}{}{}
\makeatother
% Allow footnotes in longtable head/foot
\IfFileExists{footnotehyper.sty}{\usepackage{footnotehyper}}{\usepackage{footnote}}
\makesavenoteenv{longtable}
\usepackage{graphicx}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother
\setlength{\emergencystretch}{3em} % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{5}
% table and figure caption width
\usepackage[width=0.9\textwidth]{caption}

% table formatting
\usepackage{colortbl}
\usepackage{makecell}

% lookup cross-references from the paper
\usepackage{xr}
\externaldocument{transect_paper}

% Figures and tables are prefixed with S for supplement
\renewcommand{\thetable}{S\arabic{table}}
\renewcommand{\thefigure}{S\arabic{figure}}

% put all tables and figures on their own pages
% \usepackage[nomarkers]{endfloat}
\usepackage{placeins}

% full page landscape figures
\usepackage{rotating}
% \DeclareDelayedFloatFlavour{sidewaysfigure}{figure}
\usepackage{booktabs}
\usepackage{longtable}
\usepackage{array}
\usepackage{multirow}
\usepackage{wrapfig}
\usepackage{float}
\usepackage{colortbl}
\usepackage{pdflscape}
\usepackage{tabu}
\usepackage{threeparttable}
\usepackage{threeparttablex}
\usepackage[normalem]{ulem}
\usepackage{makecell}
\usepackage{xcolor}
\ifluatex
  \usepackage{selnolig}  % disable illegal ligatures
\fi
\usepackage[style=apa,]{biblatex}
\addbibresource{all.bib}
\addbibresource{R.bib}

\title{Supplementary Information for:Long- and short-read metabarcoding technologies reveal similar spatio-temporal structures in fungal communities}
\author{Brendan Furneaux\textsuperscript{1}, \and Mohammad Bahram\textsuperscript{2,3}, \and Anna Rosling\textsuperscript{4}, \and Nourou S. Yorou\textsuperscript{5}, \and Martin Ryberg\textsuperscript{1} \and \textsuperscript{1}Program in Systematic Biology, Department of Organismal Biology, Uppsala University, \and Uppsala, Sweden \and \textsuperscript{2}Department of Ecology, Swedish University of Agricultural Sciences, Uppsala, Sweden \and \textsuperscript{3}Institute of Ecology and Earth Sciences, University of Tartu, Tartu, Estonia \and \textsuperscript{4}Program in Evolutionary Biology, Department of Ecology and Genetics, Uppsala University, \and Uppsala, Sweden \and \textsuperscript{5}Research Unit in Tropical Mycology and Plant-Fungi Interactions, LEB, \and University of Parakou, Parakou, Benin}
\date{}

\begin{document}
\maketitle

\textbf{Supplementary file 1}: Mapping between samples, indexed primers, and read accession numbers for short amplicons.

\textbf{Supplementary file 2}: Mapping between samples, indexed primers, and read accession numbers for long amplicons.

\textbf{Supplementary file 3}: ML tree of all long metabarcoding ASVs (ITS1--5.8S--ITS2--LSU) assigned to kingdom Fungi.
Thick branches represent rapid bootstrap values \textgreater90\%.
Taxonomic assignment by PHYLOTAX, based on assignments using RDPC, SINTAX, and IDTAXA, with Unite, Warcup, and RDP-LSU reference databases.
Taxon names in blue were not fully resolved with PHYLOTAX, and are given as a list of alternative assignments, along with the number of methods which gave each assignment (max 9).
Taxon names in red were reconstructed as polyphyletic.
In some cases, this may be due to a poorly identified sequence ``splitting'' the group (e.g., the single \emph{Monascus} sequence nested in \emph{Aspergillus}; Aspergillaceae and Trichocomaceae sequences nested in Onygenales), but in other cases, the tree has clearly failed to reconstruct groups which are generally considered to be monophyletic based on multi-gene trees (e.g., Dothidiomycetes, Cantharellales).

\hypertarget{supplementary-methods}{%
\section{Supplementary methods}\label{supplementary-methods}}

\hypertarget{Sampling-supplement}{%
\subsection{Sample preservation and extraction}\label{Sampling-supplement}}

Soil samples described in the main text were placed into separate 2.0~mL microtubes containing 750~mL of field lysis/preservation buffer and lysis beads (Xpedition\textsuperscript{TM} Soil/Fecal DNA miniprep, Zymo Research Corporation, Irvine, California, USA).
Due to the inavailability of refrigeration, extraction was initiated in the field according to the manufacturers instructions by lysis using a handheld bead-beater (TerraLyser\textsuperscript{TM}; Zymo Research Corporation).

An additional sample was collected at every sampling location (1-m spacing) in 2016 using LifeGuard\textsuperscript{TM} Soil Preservation Solution (MO BIO, Carlsbad, CA; USA) for preservation, without field lysis.
Sequencing results for these 50 samples differed significantly (PERMANOVA with 9999 permutations, \(p < 0.0001\), \(R^2 = 0.06\)) from samples preserved using the Xpedition\textsuperscript{TM} lysis buffer (Figures \ref{fig:buffer-pcoa}, \ref{fig:buffer-compare-long}, and \ref{fig:buffer-compare-short}); as such the LifeGuard preserved samples were excluded from our spatial analyses.
However, reads from these samples were included in the full bioinformatics workflow, including ASV calling, OTU clustering, and phylogenetic trees.

After returning to the laboratory, DNA was extracted using the Xpedition\textsuperscript{TM} Soil/Fecal Prep kit (see above).
Samples preserved using LifeGuard were first centrifuged at 10~000~g for 1~minute, after which the supernatant was removed and DNA was extracted from the remaining soil using the Soil/Fecal Prep kit as for the other samples.
DNA was quantified fluorometrically using Quant-iT\textsuperscript{TM} PicoGreen\textsuperscript{TM} dsDNA (Thermo Fisher Scientific, Waltham, MA, USA) fluorescent indicator dye on a Infinite F200 plate spectrofluorometer (Tecan Trading AG, MÃ¤nnedorf, Switzerland) according to the manufacturer's protocol.

\hypertarget{denoising-supplement}{%
\subsection{Denoising and clustering}\label{denoising-supplement}}

\hypertarget{long-numbers}{%
\subsubsection{Denoising long-reads: the numbers are against us}\label{long-numbers}}

The DADA2 algorithm requires that the seed sequence of each ASV be represented by at least two error-free reads \autocite{callahan2016}.
If sequencing errors are uniformly distributed, then the probability that a given read will be error-free is \((1-\epsilon)^L\), where \(\epsilon\) is the sequencing error rate and \(L\) is the read length in base pairs.
Then the number of reads of a given sequence that would be required to obtain two error-free reads in expectation is \(2/(1-\epsilon)^L\).
For the combination of long sequences (median \(L=1509 \text{ bp}\) after trimming) and moderate error rate (mean \(\epsilon= 0.0066\) based on ccs quality scores) for the long amplicon in this study, the expected number of reads required to achieve two error-free reads is 43 720.
Given the high diversity relative to sequencing depth in this study (485 ASVs based on PacBio short amplicons, 104 305 trimmed long amplicon reads), this requirement could not have been met for the long amplicons except by the most abundant sequences.
In comparison, the equivalent requirement for the short amplicon (\(L=265 \text{ bp}\), \(\epsilon=0.0022\)) is only 3.6 reads.

In order to bypass this limitation, we implemented a workflow which divides the long rDNA amplicon into homologous domains (new R package \texttt{LSUx}), denoises the domains individually using DADA2, and then recombines the denoised domains to form full-length denoised amplicons (new R package \texttt{tzara}).
The steps of this workflow are detailed below.

\hypertarget{splitting-the-rdna-into-regions-with-covariance-models-lsux}{%
\subsubsection{\texorpdfstring{Splitting the rDNA into regions with covariance models: \texttt{LSUx}}{Splitting the rDNA into regions with covariance models: LSUx}}\label{splitting-the-rdna-into-regions-with-covariance-models-lsux}}

Raw reads were divided into shorter regions by matching to covariance models (CM), which are similar to stochastic hidden markov models (HMM), but account for both nucleotide sequence and RNA secondary structure \autocite{eddy1994}.
First, the 5.8S rDNA was located in each read by searching for Rfam model RF0002 \autocite{kalvari2018} using \texttt{cmsearch} from Infernal 1.1.2 \autocite{nawrocki2013}, and all bases before the 5.8S were assigned to ITS1.
No attempt was made to remove the approximately 12~bp fragment of the SSU from the 5' end of ITS1 in the long amplicons; it was too short to be reliably detected by a CM or the HMMs employed by ITSx \autocite{bengtsson-palme2013}.
A reference alignment including conserved RNA base pairing between and within the 5.8S and relevant portions of LSU was generated from the fungal LSU RNA seed alignment from RDP release 11.5 \autocite{glockner2017,cole2014} by truncating after the LR5 primer site and using the reference line to annotate the variable regions \emph{sensu} \textcite{michot1984} and \textcite{raue1988}.
A CM was generated from the alignment using \texttt{cmbuild} from Infernal.
The fragment of each read beginning with the 5.8S rDNA was then aligned to the CM using \texttt{cmalign} from Infernal.
The annotation line in the CM alignment for each read was then used to split the reads into alternating more-conserved and less-conserved domains as shown in Figure~\ref{fig:regions},
where LSU1--4 represent the conserved domains of LSU flanking the variable D1--3 domains \autocite{michot1984}.
For short amplicons, only (partial) 5.8S, ITS2, and (partial) LSU1 were extracted.
Code to extract the segments, including annotated seed alignments and CMs, is available in the new R package \texttt{LSUx}, available and documented at \url{https://github.com/brendanf/LSUx}.

\hypertarget{denoising-each-domain-independently-with-dada2}{%
\subsubsection{\texorpdfstring{Denoising each domain independently with \texttt{DADA2}}{Denoising each domain independently with DADA2}}\label{denoising-each-domain-independently-with-dada2}}

Each of the extracted domains was independently filtered for length (Table~\ref{tab:region-limits}) and a maximum of three expected errors, and denoised using DADA2.
To test the performance of DADA2 on intermediate-length amplicons, adjacent domains from each read were concatenated to form full ITS (ITS1 + 5.8S + ITS2), partial LSU (LSU1 + D1 + LSU2 + D2 + LSU3 + D3 + LSU4), and partial 32S (5.8S + ITS2 + LSU) datasets, and these were also denoised.
The DADA2 error model was fit using the 5.8S region for long amplicons, and using the entire read for short amplicons.
Independent error models were fit for each sequencing strategy (i.e., long \emph{vs.} short amplicons, different sequencing technologies).
DADA2 was run with complete pooling for PacBio libraries, and pseudo-pooling
for Ion Torrent libraries.
Chimeras within each region were removed using \texttt{removeBimeraDenovoTable} from DADA2.

\hypertarget{generation-of-full-length-denoised-asvs-tzara}{%
\subsubsection{\texorpdfstring{Generation of full-length denoised ASVs: \texttt{tzara}}{Generation of full-length denoised ASVs: tzara}}\label{generation-of-full-length-denoised-asvs-tzara}}

For each raw read, the denoised sequences of the different domains were concatenated to form a set of full-length sequences.
For reads which were not assigned a denoised sequence for each domain, the raw read for the domain was used instead.
Because ITS2 is the most variable of the amplified domains as indicated by the greatest number of unique ASVs (Figure~\ref{fig:regions}b), reads with identical ITS2 regions are expected to have highly similar sequences in the other regions, unless the amplicon was chimeric.
Therefore, concatenated sequences with identical ITS2 were clustered, and each cluster was aligned in R using the \texttt{DECIPHER} package \autocite{wright2015}.
Outlier sequences, as determined by mean pairwise distance from the rest of the alignment, were removed from each alignment using the \texttt{odseq} package \autocite{jehl2015}, using the default threshold of 0.025.
The consensus of the remaining aligned sequences was assigned as the full-length ASV sequence for the entire cluster.
Full-length ASV sequences with more than three ambiguous bases (i.e., no nucleotide \textgreater50\% at a given position) were removed.
The count and sample distribution of reads assigned to each full-length ASV were calculated in order to form a sample Ã ASV community matrix.
A similar process was used to generate a consensus ITS (ITS1--5.8S--ITS2) and LSU (LSU1--D1--LSU2--D2--LSU3--D3--LSU4) sequence for each ASV.
The process of assigning consensus full-length ASVs was carried out using the new \texttt{tzara} package for R, available and documented at \url{https://github.com/brendanf/tzara}.

The two packages are modular; \texttt{LSUx} can be used to extract domains from LSU for any purpose, and \texttt{tzara} can also be used to recombine denoised sequences which were originally separated using other approaches, most notably using the popular and well-established ITSx \autocite{bengtsson-palme2013}.
However, for our dataset, the single LSU region which would be extracted by ITSx was still too long to be effectively denoised by DADA2 (Figure~\ref{fig:regions}b in main text).

\hypertarget{phylogeny-supplement}{%
\subsection{Phylogenetic inference and taxonomy assignment}\label{phylogeny-supplement}}

\hypertarget{reannotate}{%
\subsubsection{Uniform taxonomic classification for Unite, Warcup, and RDP-LSU}\label{reannotate}}

The taxonomic classification system is based on the Unite database version 8 \autocite{nilsson2019a}.
In particular, the classification for fungi was according to \textcite{tedersoo2018}, and for non-fungal eukaryotes was according to the proposed system of \textcite{tedersoo2017c} as described in \autocite{tedersoodata2017}.
Although the latter system is not formally published, it is consistent with the annotations for non-fungal eukaryotes in the Unite database.
Additionally, it is a system with both purportedly monophyletic taxa and a uniform set of taxon ranks, which make it more appropriate for sequence-based taxonomic assignment algorithms than more accepted classification systems such as that of the International Society of Protistologists \autocite{adl2019}, which utilizes hierarchical nameless ranks.
FASTA-format files for the RDP-LSU, Warcup, and Unite reference databases with unified classifications, as well as scripts used to generate them, are available at \url{https://github.com/brendanf/reannotate}.

\hypertarget{rooting}{%
\subsubsection{Rooting the fungal tree}\label{rooting}}

The phylogenetic tree was rooted based on sequences which were confidently assigned to kingdom based on the primary taxonomic assignments.
Assignments based on Warcup were not used at this step because non-Fungi are not included in the dataset, and thus it cannot be used to directly assign kingdoms.
Assignments were considered confident if at least five of the six remaining primary assignment methods (2 databases Ã 3 algorithms) resulted in the same kingdom-level assignment, with no conflicting assignments.
Two ASVs belonging to \emph{Tulasnella}, which is known to have an accelerated rate of molecular evolution in the rDNA and often forms long, unstable branches in phylogenies \autocite{moncalvo2006}, were excluded at this step.
This resulted in 516 confident kingdom assignments, with 41 assigned to Alveolata, 453 assigned to Fungi, 1 assigned to Metazoa, and 21 assigned to Viridiplantae.
Inspection of the tree verified that these four kingdoms were reciprocally monophyletic, and the clade containing Alveolata and Viridiplantae was used as an outgroup to root the tree.
The kingdom Fungi was then identified as the minimal clade containing all ASVs which were confidently identified as kingdom Fungi.
ASVs falling outside this clade, including the excluded \emph{Tulasnella} sequences, were not included in downstream fungal community analysis.

\hypertarget{phylotax}{%
\subsubsection{\texorpdfstring{Refining taxonomic assignments with a phylogenetic tree: \texttt{PHYLOTAX}}{Refining taxonomic assignments with a phylogenetic tree: PHYLOTAX}}\label{phylotax}}

PHYLOTAX refinements are based on (1) taxonomic assignments from other, primary assignment algorithms, and (2) a phylogenetic tree.
Taxa at each rank are assigned to a node and all its descendants if that taxon is consistent with the reference-based taxonomic assignments for each of the descendants (Figure~\ref{fig:phylotax}).
A taxon assignment is considered to be consistent if at least one primary algorithm assigned that taxon at greater than 50\% confidence, or if no primary algorithm successfully classified the sequence at greater than 50\% confidence.
The result of this process is twofold.
First, it gives a taxonomic assignment to ASVs which were previously unassigned if they are nested within a clade with consistent preliminary assignments.
Second, it resolves conflicts between the different reference-based algorithms, but only if one of the alternatives is consistent with the assignments of other ASVs in the same clade.
PHYLOTAX is implemented in the new R package \texttt{phylotax} at \url{https://github.com/brendanf/phylotax}.

\hypertarget{sequencing-strategy-supplement}{%
\subsection{Effect of sequencing strategy on recovered community}\label{sequencing-strategy-supplement}}

ASV and OTU accumulation curves were generated using the \texttt{iNEXT} package \autocite{hsieh2016}.
Additionally, ASV and OTU richness was estimated for each sample by rarefying to 100 sequences per sample and taking the average of 100 repetitions.
ASV and OTU richness were then compared for each biological sample as sequenced by the different strategies.
For each pair of strategies, the best fit slope for the ASV richness recovered by each strategy was calculated by total least squares using the \texttt{deming} package \autocite{R-deming}, with the intercept constrained to be 0.
For this analysis, samples in the PacBio and Illumina datasets were combined bioinformatically to match the corresponding samples from the incompletely demultiplexed Ion Torrent dataset.

The effect of sequencing technology and amplicon length on the recovered fungal community
composition, as assessed by the Bray-Curtis dissimilarity was determined by PERMANOVA \autocite{anderson2003} using the \texttt{adonis2} function in the \texttt{vegan} package \autocite{R-vegan}.
Comparisons were made between the fungal communities recovered by the three sequencing strategies that were successfully demultiplexed (Illumina short, PacBio short, PacBio long).

In order to detect potential biases at larger taxonomic scales, ASVs were clustered according to the assigned taxonomic class.
Only samples where all three strategies yielded at least 100~fungal reads ( samples), and classes which represented at least 1\% of reads in at least one sample ( classes), were included.
The community matrix was normalized to proportional abundance by dividing the counts for each class in each sample by the total number of reads for that sample, but were not rarefied to even sampling depth \autocite{mcmurdie2014}.
PERMANOVA included three terms: an indicator for soil sample, comprising all spatiotemporal effects; amplicon length (long vs.~short); and sequencing technology (Illumina vs.~PacBio).
Partial Principal Coordinates Analysis (PPCoA) was applied to the same dissimilarity matrix using the \texttt{capscale} function in \texttt{vegan} \autocite{R-vegan}.
Spatiotemporal effects, i.e.~differences between different samples, were partialed out using a \texttt{Condition()} term in the \texttt{capscale} model in order to visualize effects due to sequencing technology and amplicon length.

A similar analysis was also applied to only fungi classified as ECM, clustered at the family level.
Sequences were assigned as ECM based on taxonomic assignments using the FUNGuild database \autocite{nguyen2016funguild} via the R package \texttt{FUNGuildR} (\url{https://github.com/brendanf/FUNGuildR}).
All taxa which included ``Ectomycorrhiza'' in the guild assignment at any level of confidence were included.

Overall community composition, and differences in composition between amplicon libraries, were visualized as heat trees using the \texttt{metacoder} package \autocite{R-metacoder}.

\hypertarget{spatiotemporal-supplement}{%
\subsection{Spatiotemporal analysis}\label{spatiotemporal-supplement}}

Empirical distance-decay curves were generated by plotting mean community dissimilarity as a function of spatial distance, and fit to an exponential model of the form given by Legendre and Legendre (2012) using the nls function in R.
Points in the empirical distance-decay curve were weighted by the number of comparisons within the distance class and the inverse of the distance for the purposes of model fitting.
For datasets where the Mantel correlogram indicated spatial correlation between samples taken in separate years, the model was re-fit with an additional term to represent temporal correlation:
\[D = C_0 + C_1\left[1 - \exp\left(-3  \left(\frac{d}{a_d} + \frac{t}{a_t}\right)\right)\right]\]
where \(D\), \(d\), and \(t\) represent the community dissimilarity, spatial distance, and time lag between samples, respectively, and the parameters are \(C_0\), the community dissimilarity between replicate samples (``nugget''); \(C_0 + C_1\), the community dissimilarity at long distances (``sill''); \(a_d\) the spatial range at which the community dissimilarity has moved 95\% of the way from ``nugget'' to ``sill''; and \(a_t\), the equivalent temporal range.
The 95\% confidence intervals were calculated for the spatial and temporal range parameters by profiling using the \texttt{MASS} package.

\newpage

\hypertarget{supplementary-figures}{%
\section{Supplementary Figures}\label{supplementary-figures}}

\listoffigures






\begin{figure}

{\centering \includegraphics[width=1\linewidth]{transect_supplement_files/figure-latex/buffer-pcoa-1} 

}

\caption[Unconstrained Principal Coordinate Analysis (PCoA) ordination of samples preserved with different buffers]{Unconstrained Principal Coordinate Analysis (PCoA) ordination of samples preserved with different buffers, based on Bray-Curtis dissimilarities of relative read abundances.
For each sequencing strategy, only those sampling locations with more than 100 reads each for all three year/preservation buffer combinations were included.}\label{fig:buffer-pcoa}
\end{figure}











\begin{figure}

{\centering \includegraphics[width=1\linewidth]{transect_supplement_files/figure-latex/buffer-compare-long-1} 

}

\caption[Comparison of read counts by taxonomic groups for different preservation buffers (Xpedition vs.~LifeGuard) and sampling year, for long amplicon libraries]{Comparison of read counts by taxonomic groups for different preservation buffers (Xpedition vs.~LifeGuard) and sampling year, for long amplicon libraries. Reference heat tree (bottom left) shows labeled taxonomic tree.
Nodes and branches are scaled to represent mean relative read abundance across samples.
Colored comparison trees (top and right) illustrate variation in relative read abundance between samples preserved by different techniques or from different years.
For each comparison tree, colors indicate the base 10 log relative abundance ratio for each taxon between two datasets, with red-orange colors indicating greater relative abundance in the dataset indicated above the tree (Xpedition 2015, LifeGuard 2016), and blue-green colors indicating greater relative abundance in the dataset indicated to the right of the tree (Xpedition 2016, Xpedition 2015).
The two Xpedition datasets (blue-green colors in top right and center right trees) show consistently greater relative abundance of Basidiomycota, especially Russulales and Sebacinales, as well as Glomeromycota, Morteriellomycota, Geoglossomycetes, and Pezizomycetes,
while the LifeGuard samples (red-orange colors in top right and center right trees) showed greater relative abundance of Ascomycota in general, especially Aspergillaceae, and also Geminibasidiomycetes.
Differences between the two Xpedition datasets (top center tree) are less pronounced.}\label{fig:buffer-compare-long}
\end{figure}











\begin{figure}

{\centering \includegraphics[width=1\linewidth]{transect_supplement_files/figure-latex/buffer-compare-short-1} 

}

\caption[Comparison of read counts by taxonomic groups for different preservation buffers (Xpedition vs.~LifeGuard) and sampling year, for short amplicon libraries]{Comparison of read counts by taxonomic groups for different preservation buffers (Xpedition vs.~LifeGuard) and sampling year, for short amplicon libraries. Reference heat tree (bottom left) shows labeled taxonomic tree.
Nodes and branches are scaled to represent mean relative read abundance across samples.
Colored comparison trees (top and right) illustrate variation in relative read abundance between samples preserved by different techniques or from different years.
For each comparison tree, colors indicate the base 10 log relative abundance ratio for each taxon between two datasets, with red-orange colors indicating greater relative abundance in the dataset indicated above the tree (Xpedition 2015, LifeGuard 2016), and blue-green colors indicating greater relative abundance in the dataset indicated to the right of the tree (Xpedition 2016, Xpedition 2015).
The two Xpedition datasets (blue-green colors in top right and center right trees) show consistently greater relative abundance of Basidiomycota, especially Sebacinales, as well as Glomeromycota, Geoglossomycetes, and Pezizomycetes,
while the LifeGuard samples (red-orange colors in top right and center right trees) showed greater relative abundance of Ascomycota in general, especially Aspergillaceae, as well as Geminibasidiomycetes.
Differences between the two Xpedition datasets (top center tree) are less pronounced.}\label{fig:buffer-compare-short}
\end{figure}









\begin{figure}

{\centering \includegraphics[width=1\linewidth]{transect_supplement_files/figure-latex/phylotax-1} 

}

\caption[Phylogenetic refinement of taxonomic assignments (PHYLOTAX)]{Phylogenetic refinement of taxonomic assignments (PHYLOTAX). In the example, a clade includes seven OTUs (A-F), which have been identified by two methods as belonging to taxa ``Tax1'' or ``Tax2'', or are unidentified (``unk''), as shown in the tip labels.
No taxon is assigned at node~1, because one of the child branches (A) is completely unassigned.
No taxon is assigned at node~2, because the assignments at C and F are inconsistent.
Node~3 is assigned to Tax2 because this is consistent with at least one of the assignments for both B and C.
Node~4 (and thus also node~5) is assigned to Tax1 because this is consistent with at least one of the assignments for D and F, and because E, which is nested within the clade defined by D and F, is completely unassigned.}\label{fig:phylotax}
\end{figure}








\begin{sidewaysfigure}

{\centering \includegraphics[width=1\linewidth]{transect_supplement_files/figure-latex/sample-depth-1} 

}

\caption[DNA concentrations after extraction and PCR, and sequencing depth along transects at the two sites Ang and Gan for the years 2015 and 2016]{DNA concentrations after extraction (leftmost column) and PCR (second and third columns), and sequencing depth (right three columns) along transects at the two sites Ang and Gan for the years 2015 and 2016.
Sequencing reads which were filtered out during quality control are colored dark grey.
Dotted horizontal line at 100~reads indicates cutoff for inclusion in community analysis.
2016 samples were preserved using two different methods (LifeGuard, Xpedition), while 2015 samples were preserved only using Xpedition.}\label{fig:sample-depth}
\end{sidewaysfigure}





\begin{figure}

{\centering \includegraphics[width=1\linewidth]{transect_supplement_files/figure-latex/qual-check-1} 

}

\caption[Read quality profiles]{Read quality profiles, based on quality scores assigned by technology-specific basecallers. \textbf{(a)} per-base error rate for different sequencing strategies; \textbf{(b)} expected number of errors per read for different sequencing strategies; \textbf{(c)} expected number of errors per read for regions extracted from PacBio long amplicon reads. In \textbf{(a)} and \textbf{(b)}, R1 and R2 denote the forward and reverse reads, respectively, from Illumina paired-end sequencing. PacBio short, Ion Torrent, and Illumina datasets were all sequenced from the same short amplicon library. In \textbf{(b)} and \textbf{(c)}, quality filtering threshold of 3 expected errors marked as a dashed vertical line.}\label{fig:qual-check}
\end{figure}





\begin{figure}

{\centering \includegraphics[width=1\linewidth]{transect_supplement_files/figure-latex/full-length-1} 

}

\caption[Comparison of the length of denoised sequences from different length amplicons and sequencing technologies]{Comparison of the length of denoised sequences from different length amplicons and sequencing technologies. Length distribution (\emph{a}, \emph{c}) and empirical cumulative distribution function (ECDF; \emph{b}, \emph{d}) for short (\emph{a}, \emph{b}) and long (\emph{c}, \emph{d}) amplicons, respectively.}\label{fig:full-length}
\end{figure}





\begin{figure}

{\centering \includegraphics[width=1\linewidth]{transect_supplement_files/figure-latex/region-lengths-1} 

}

\caption[Length distribution for each region extracted from long amplicons]{Length distribution for each region extracted from long amplicons. Filtering limits are shown as dashed vertical lines.}\label{fig:region-lengths}
\end{figure}





\begin{figure}

{\centering \includegraphics[width=1\linewidth]{transect_supplement_files/figure-latex/ITS2-length-1} 

}

\caption[Comparison of the length of denoised ITS2 sequences extracted from different length amplicons and sequencing technologies]{Comparison of the length of denoised ITS2 sequences extracted from different length amplicons and sequencing technologies. \emph{a}) Length distribution. \emph{b}) Empirical cumulative distribution function.}\label{fig:ITS2-length}
\end{figure}





\begin{figure}

{\centering \includegraphics[width=1\linewidth]{transect_supplement_files/figure-latex/read-compare-1} 

}

\caption[Comparison between read numbers for different sequencing strategies]{Comparison between read numbers for different sequencing strategies, by ASV (\textbf{a}) and 97\% OTU (\textbf{b}). ASVs/OTUs which were detected by one sequencing strategy but not the other are plotted as tick marks along the axes. Dashed line represents a constant ratio of read numbers. The blue line is a LOESS smooth of the data, with associated uncertainty in grey shading. \(R^2\) value displayed is for log-transformed non-zero read numbers.}\label{fig:read-compare}
\end{figure}








\begin{figure}

{\centering \includegraphics[width=1\linewidth]{transect_supplement_files/figure-latex/accum-OTU-1} 

}

\caption[97\% OTU accumulation curves]{97\% OTU accumulation curves. Each curve represents rarefaction of a single sample or pooled pair of samples.
Points at the end of each curve represent the actual read depth and observed OTU richness.
Panel at left has enlarged scale to show PacBio more clearly.
Vertical dashed line at 100 reads indicates rarefaction level for OTU richness comparisons.}\label{fig:accum-OTU}
\end{figure}





\begin{figure}

{\centering \includegraphics[width=1\linewidth]{transect_supplement_files/figure-latex/alpha-compare-otu-1} 

}

\caption[Comparison of OTU richness between sequencing technologies]{Comparison of OTU richness between sequencing technologies Each point represents the richness of one or two pooled samples, as determined by two different sequencing strategies. All values represent the average of 100 replicate rarefactions with a sample depth of 100 reads. Because samples in the same well on different plates could not be demultiplexed in the Ion Torent dataset, these samples were also bioinformatically pooled in the other datasets prior to rarefaction. Blue lines are total least squares fit with 95\% confidence interval, with the given slope (and 95\% confidence interval) and R\textsuperscript{2} value. Dashed diagonal line indicates ideal slope of 1.}\label{fig:alpha-compare-otu}
\end{figure}












\begin{sidewaysfigure}

{\centering \includegraphics[width=1\linewidth]{transect_supplement_files/figure-latex/taxa-1} 

}

\caption[Heat tree summarizing taxonomic composition of soil community]{Heat tree summarizing taxonomic composition of soil community. Color and size of nodes represent the fractional ASV richness of that taxonomic group.
Color and thickness of branches represent the fractional read abundance of all ASVs belonging to that taxonomic group.
Diverse but relatively rare groups are thus shown as large, dark-colored nodes on relatively thin, light-colored branches (e.g., Dothideomycetes), while common but relatively non-diverse groups are shown as small, light-colored nodes on relatively thick, dark-colored branches (e.g., Russulales).
Groups which are not represented by at least 1\% of reads or 1\% of ASVs in any dataset are collapsed into nodes labeled "*``.
ASVs which could not be further assigned are labelled''?".
ASV richness and read abundance are displayed as the mean across sequencing runs.
Taxonomic assignment is by strict consensus of multiple ITS databases (Unite, Warcup) and assignment algorithms (RDPC, SINTAX, IDTAXA).
The RDP-LSU database was not included, to ensure consistent assignment between long and short amplicon libraries.}\label{fig:taxa}
\end{sidewaysfigure}











\begin{figure}

{\centering \includegraphics[width=1\linewidth]{transect_supplement_files/figure-latex/heattree-length-compare-1} 

}

\caption[Variation in read abundance and diversity for different amplicons across the taxonomic tree]{Variation in read abundance and diversity for different amplicons across the taxonomic tree. Color of nodes indicates variation in read abundance, while color of edges represents variation in ASV richness.
In both cases red-orange represents increased prevalence in the short amplicon library, while blue-green represents increased prevalence in the long amplicon library.
Values are log\textsubscript{10} of the ratio of mean values across sequencing runs for each amplicon.
Groups which are not represented by at least 1\% of reads or 1\% of ASVs in any dataset are collapsed into nodes labeled "*``.
ASVs which could not be further assigned are labelled''?".
Only nodes with a log read abundance ratio greater than 0.5 (abundance ratio \textgreater{} 3.2) are labeled.
Refer to Figure~\ref{fig:taxa} for complete taxon labeling.}\label{fig:heattree-length-compare}
\end{figure}





\begin{figure}

{\centering \includegraphics[width=1\linewidth]{transect_supplement_files/figure-latex/short-its-1} 

}

\caption[Kingdom-level taxonomic composition of ITS2 ASVs \(\le 140\) bp]{Kingdom-level taxonomic composition of ITS2 ASVs \(\le 140\) bp. Vertical axis shows fraction of total reads \(\le 140\) bp in each dataset.}\label{fig:short-its}
\end{figure}








\begin{figure}

{\centering \includegraphics[width=1\linewidth]{transect_supplement_files/figure-latex/ecm-1} 

}

\caption[Assignment of ECM status to fungal ASVs using FUNGuild database and taxonomic assignments]{Assignment of ECM status to fungal ASVs using FUNGuild database and taxonomic assignments. Reads which were not identified to kingdom, or which were identified as a non-Fungi kingdom, are not included.
``unidentified'' denotes ASVs which were assigned to kingdom Fungi, but could not be identified to family level or could not be assigned ECM status based on the available identification.
Height of each bar represents the fraction of reads represented by that bar.
Results are shown for different sequencing technologies, amplicons (Long, Short), reference databases (Unite, Warcup, RDP-LSU; ``All'' denotes algorithms which combine results from multiple databases), and taxonomic assignment algorithms.}\label{fig:ecm}
\end{figure}










\begin{figure}

{\centering \includegraphics[width=1\linewidth]{transect_supplement_files/figure-latex/ecm-heattree-1} 

}

\caption[Heat tree showing variation in read abundance and diversity for different amplicons, for ECM taxa only]{Heat tree showing variation in read abundance and diversity for different amplicons, for ECM taxa only. Red-orange edges (nodes) represent increased read (ASV) count for the short amplicon library, while blue-green edges (nodes) represent increased read (ASV) count for the long amplicon library.
Values are log\textsubscript{10} of the ratio of mean values across all sequencing technologies for each amplicon.
Groups which are not represented by at least 1\% of reads or 1\% of ASVs in any dataset are collapsed into nodes labeled "*``.
ASVs which could not be further assigned are labelled''?".
Taxonomic assignment is by strict consensus of multiple ITS databases (Unite, Warcup) and assignment algorithms (RDPC, SINTAX, IDTAXA).
The RDP-LSU database was not included, to ensure consistent assignment between long and short amplicon libraries.}\label{fig:ecm-heattree}
\end{figure}









\begin{figure}

{\centering \includegraphics[width=1\linewidth]{transect_supplement_files/figure-latex/ppcoa-1} 

}

\caption[Partial Principal Coordinates Analysis (PPCoA) of fungal community using different metabarcoding strategies]{Partial Principal Coordinates Analysis (PPCoA) of fungal community using different metabarcoding strategies, for all fungi (\textbf{a}) and ECM fungi (\textbf{b}).
Ordination based on the residual variation after spatio-temporal variation (i.e., variation between different soil samples) has been accounted for.
Only samples with at least 100 reads for all three strategies are included.
ASVs are clustered by taxonomic class for all fungi, and by family for ECM fungi.
Abbreviations represent the approximate direction of increase for the classes and families with the highest scores: \textbf{So}: Sordariomycetes, \textbf{Ag}: Agaricomycetes, \textbf{Pe}: Pezizomycetes, \textbf{Eu}: Eurotiomycetes, and \textbf{Do}: Dothideomycetes; \textbf{In}: Inocybaceae, \textbf{Am}: Amanitaceae, \textbf{Sc}: Sclerodermataceae, \textbf{Ru}: Russulaceae, \textbf{Pe}: Pezizaceae, and \textbf{Cl}: Clavulinaceae.}\label{fig:ppcoa}
\end{figure}





\begin{figure}

{\centering \includegraphics[width=1\linewidth]{transect_supplement_files/figure-latex/correlog-1} 

}

\caption[Mantel correlograms for community dissimilarities and spatio-temporal distance]{Mantel correlograms for community dissimilarities and spatio-temporal distance. Community dissimilarities are Bray-Curtis (top three rows) or weighted Unifrac (bottom row) dissimilarities and spatial distance, for short (top two rows) and long (bottom two rows) amplicons, and for the entire soil fungal community (left) or only ECM taxa (right). Unifrac distance was not calculated for short amplicons because it requires that the sequences are placed on a phylogenetic tree.}\label{fig:correlog}
\end{figure}

\FloatBarrier

\hypertarget{supplementary-tables}{%
\section{Supplementary tables}\label{supplementary-tables}}

\listoftables

\newpage

\FloatBarrier

\begin{table}

\caption{\label{tab:region-limits}Minimum and maximum allowed lengths for each extracted region.}
\centering
\begin{tabular}[t]{lrr}
\toprule
Region & Min. length & Max. length\\
\midrule
ITS1 & 50 & 500\\
5.8S & 50 & 200\\
ITS2 & 50 & 500\\
LSU1 & 25 & 200\\
D1 & 20 & 500\\
\addlinespace
LSU2 & 50 & 500\\
D2 & 20 & 2999\\
LSU3 & 30 & 60\\
D3 & 50 & 500\\
LSU4 & 50 & 500\\
\bottomrule
\end{tabular}
\end{table}












\begin{table}

\caption[Number of Amplicon Sequence Variants (ASV) and reads at different pipeline stages]{\label{tab:bioinfo}Number of Amplicon Sequence Variants (ASV) and reads at different pipeline stages. \textbf{Raw}: raw reads as delivered by sequencing center (CCS reads for PacBio);
\textbf{Trim}: reads with primers and demultiplexing barcodes removed;
\textbf{Filter (full)}: full length amplicons with a maximum of three expected errors;
\textbf{LSUx}: reads with a positive CM hit for 5.8S, allowing regions to be extracted;
\textbf{Filter (ITS2)}: extracted ITS2 regions with a maximum of three expected errors;
\textbf{ITS2}: ASVs obtained for the ITS2 region, and reads successfully mapped to an ITS2 ASV;
\textbf{short}, \textbf{ITS}, \textbf{LSU}, \textbf{long}: ITS2-based ASVs mapped to consensus ASVs for longer regions, where "short" and "long" denote the full-length short and long amplicons, respectively.
\emph{Italicized} numbers represent mapping via ITS2 ASVs to amplicons originally sequenced by another technology.}
\centering
\begin{tabular}[t]{lrrrrrrrr}
\toprule
\multicolumn{1}{c}{ } & \multicolumn{4}{c}{PacBio RS II} & \multicolumn{2}{c}{Ion Torrent Ion S5} & \multicolumn{2}{c}{Illumina MiSeq} \\
\cmidrule(l{3pt}r{3pt}){2-5} \cmidrule(l{3pt}r{3pt}){6-7} \cmidrule(l{3pt}r{3pt}){8-9}
\multicolumn{1}{c}{ } & \multicolumn{2}{c}{Long} & \multicolumn{2}{c}{Short} & \multicolumn{2}{c}{Short} & \multicolumn{2}{c}{Short} \\
\cmidrule(l{3pt}r{3pt}){2-3} \cmidrule(l{3pt}r{3pt}){4-5} \cmidrule(l{3pt}r{3pt}){6-7} \cmidrule(l{3pt}r{3pt}){8-9}
  & ASVs & reads & ASVs & reads & ASVs & reads & ASVs & reads\\
\midrule
Raw & â & 125,034 & â & 49,511 & â & 20,717,742 & â & 10,756,939\\
Trim & â & 104,305 & â & 41,096 & â & 15,208,677 & â & 9,513,433\\
Filter (full) & â & 44,190 & â & 38,742 & â & 12,177,705 & â & 7,674,712\\
\addlinespace
LSUx & â & 100,235 & â & 40,137 & â & 14,946,121 & â & â\\
Filter (ITS2) & â & 87,860 & â & 38,756 & â & 13,474,693 & â & â\\
ITS2 & 978 & 82,460 & 564 & 37,218 & 12,964 & 13,416,070 & 4,478 & 6,763,654\\
\addlinespace
short & \em{598} & \em{65,191} & 562 & 37,214 & 12,246 & 13,413,990 & 4,419 & 6,753,567\\
ITS & 806 & 81,234 & \em{268} & \em{30,582} & \em{470} & \em{7,938,390} & \em{491} & \em{4,848,591}\\
LSU & 779 & 81,096 & \em{268} & \em{30,741} & \em{479} & \em{8,074,310} & \em{498} & \em{4,930,083}\\
long & 708 & 80,290 & \em{257} & \em{30,481} & \em{434} & \em{7,895,829} & \em{457} & \em{4,817,606}\\
\bottomrule
\end{tabular}
\end{table}

\setlength{\tabcolsep}{1ex}





\begin{table}

\caption[Correspondences between ASVs found by different sequencing strategies]{\label{tab:venn-table-asv}Correspondences between ASVs found by different sequencing strategies. Each row shows the number of ASVs (\emph{ASVs}) shared uniquely by one or more datasets, including the fraction of total ASVs for each dataset (\emph{ASVs frac}), the number of reads represented by those ASVs in each dataset (\emph{reads}) and the fraction of total reads for the dataset (\emph{reads frac}).}
\centering
\begin{tabular}[t]{rcrrcrrcrrcrrr}
\toprule
\multicolumn{2}{c}{ } & \multicolumn{6}{c}{PacBio RS II} & \multicolumn{3}{c}{Illumina MiSeq} & \multicolumn{3}{c}{Ion Torrent Ion S5} \\
\cmidrule(l{3pt}r{3pt}){3-8} \cmidrule(l{3pt}r{3pt}){9-11} \cmidrule(l{3pt}r{3pt}){12-14}
\multicolumn{2}{c}{ } & \multicolumn{3}{c}{Long} & \multicolumn{3}{c}{Short} & \multicolumn{3}{c}{Short} & \multicolumn{3}{c}{Short} \\
\cmidrule(l{3pt}r{3pt}){3-5} \cmidrule(l{3pt}r{3pt}){6-8} \cmidrule(l{3pt}r{3pt}){9-11} \cmidrule(l{3pt}r{3pt}){12-14}
\makecell[c]{\\} & ASVs & \makecell[c]{ASVs\\frac} & reads & \makecell[c]{reads\\frac} & \makecell[c]{ASVs\\frac} & reads & \makecell[c]{reads\\frac} & \makecell[c]{ASVs\\frac} & reads & \makecell[c]{reads\\frac} & \makecell[c]{ASVs\\frac} & reads & \makecell[c]{reads\\frac}\\
\midrule
 & 378 & 0.39 & 17k & 0.21 & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â }\\
 & 14 & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & 0.02 & 232 & 0.01 & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â }\\
 & 868 & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & 0.19 & 165k & 0.02 & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â }\\
 & 9378 & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & 0.72 & 1.8M & 0.14\\
\addlinespace
 & 28 & 0.03 & 682 & 0.01 & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & 0.01 & 12k & 0.00 & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â }\\
 & 40 & 0.04 & 879 & 0.01 & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & 0.00 & 3.2k & 0.00\\
 & 26 & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & 0.05 & 420 & 0.01 & 0.01 & 29k & 0.00 & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â }\\
 & 10 & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & 0.02 & 148 & 0.00 & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & 0.00 & 2.4k & 0.00\\
 & 2791 & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & 0.62 & 896k & 0.13 & 0.22 & 2.1M & 0.16\\
\addlinespace
 & 20 & 0.02 & 1.1k & 0.01 & 0.04 & 1.3k & 0.03 & 0.00 & 149k & 0.02 & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â }\\
 & 251 & 0.26 & 9k & 0.11 & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & 0.06 & 316k & 0.05 & 0.02 & 611k & 0.05\\
 & 233 & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & 0.41 & 5.5k & 0.15 & 0.05 & 703k & 0.10 & 0.02 & 1.4M & 0.10\\
\addlinespace
 & 261 & 0.27 & 54k & 0.65 & 0.46 & 30k & 0.80 & 0.06 & 4.5M & 0.66 & 0.02 & 7.5M & 0.56\\
\midrule
Total & 14298 & 1.00 & 82k & 1.00 & 1.00 & 37k & 1.00 & 1.00 & 6.8M & 1.00 & 1.00 & 13M & 1.00\\
\bottomrule
\end{tabular}
\end{table}





\begin{table}

\caption[Correspondences between 97\% OTUs found by different sequencing strategies]{\label{tab:venn-table-otu}Correspondences between 97\% OTUs found by different sequencing strategies. Each row shows the number of OTUs (\emph{OTUs}) shared uniquely by one or more datasets, including the fraction of total OTUs for each dataset (\emph{OTUs frac}), the number of reads represented by those OTUs in each dataset (\emph{reads}) and the fraction of total reads for the dataset (\emph{reads frac}).}
\centering
\begin{tabular}[t]{rcrrcrrcrrcrrr}
\toprule
\multicolumn{2}{c}{ } & \multicolumn{6}{c}{PacBio RS II} & \multicolumn{3}{c}{Illumina MiSeq} & \multicolumn{3}{c}{Ion Torrent Ion S5} \\
\cmidrule(l{3pt}r{3pt}){3-8} \cmidrule(l{3pt}r{3pt}){9-11} \cmidrule(l{3pt}r{3pt}){12-14}
\multicolumn{2}{c}{ } & \multicolumn{3}{c}{Long} & \multicolumn{3}{c}{Short} & \multicolumn{3}{c}{Short} & \multicolumn{3}{c}{Short} \\
\cmidrule(l{3pt}r{3pt}){3-5} \cmidrule(l{3pt}r{3pt}){6-8} \cmidrule(l{3pt}r{3pt}){9-11} \cmidrule(l{3pt}r{3pt}){12-14}
\makecell[c]{\\} & OTUs & \makecell[c]{OTUs\\frac} & reads & \makecell[c]{reads\\frac} & \makecell[c]{OTUs\\frac} & reads & \makecell[c]{reads\\frac} & \makecell[c]{OTUs\\frac} & reads & \makecell[c]{reads\\frac} & \makecell[c]{OTUs\\frac} & reads & \makecell[c]{reads\\frac}\\
\midrule
 & 326 & 0.35 & 16k & 0.20 & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â }\\
 & 1 & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & 0.00 & 2 & 0.00 & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â }\\
 & 133 & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & 0.05 & 8.1k & 0.00 & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â }\\
 & 4054 & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & 0.60 & 192k & 0.01\\
\addlinespace
 & 4 & 0.00 & 56 & 0.00 & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & 0.00 & 935 & 0.00 & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â }\\
 & 45 & 0.05 & 950 & 0.01 & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & 0.01 & 2.6k & 0.00\\
 & 2 & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & 0.00 & 18 & 0.00 & 0.00 & 3.1k & 0.00 & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â }\\
 & 4 & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & 0.01 & 18 & 0.00 & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & 0.00 & 712 & 0.00\\
 & 1866 & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & 0.67 & 332k & 0.05 & 0.28 & 919k & 0.07\\
\addlinespace
 & 242 & 0.26 & 5.1k & 0.06 & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & 0.09 & 124k & 0.02 & 0.04 & 350k & 0.03\\
 & 221 & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & \cellcolor{gray}{Â Â } & 0.42 & 4.2k & 0.11 & 0.08 & 629k & 0.09 & 0.03 & 1.3M & 0.10\\
\addlinespace
 & 303 & 0.33 & 60k & 0.73 & 0.57 & 33k & 0.89 & 0.11 & 5.7M & 0.84 & 0.04 & 11M & 0.79\\
\midrule
Total & 7201 & 1.00 & 82k & 1.00 & 1.00 & 37k & 1.00 & 1.00 & 6.8M & 1.00 & 1.00 & 13M & 1.00\\
\bottomrule
\end{tabular}
\end{table}



\begin{table}

\caption{\label{tab:variofit}Parameters for exponential distance-decay fits. Range: range at which exponential function is at 95\% of its asymptotic value.}
\centering
\begin{tabular}[t]{lllllll}
\toprule
Guild & Amplicon & Tech & Algorithm & Metric & Space Range (m) & Time Range (a)\\
\midrule
All Fungi & Short & Illumina & Cons & Bray-Curtis & 13 (7.4â 23) & --\\
All Fungi & Short & Illumina & PHYLO & Bray-Curtis & 13 (7.4â 23) & --\\
All Fungi & Short & PacBio & Cons & Bray-Curtis & 31 ( 15â 57) & --\\
All Fungi & Short & PacBio & PHYLO & Bray-Curtis & 31 ( 15â 57) & --\\
All Fungi & Long & PacBio & PHYLO & Bray-Curtis & 18 (  9â 41) & 3.3 (1.9â6.9)\\
All Fungi & Long & PacBio & PHYLO & W-UniFrac & 466 (0â5471) & --\\
\addlinespace
ECM & Short & Illumina & Cons & Bray-Curtis & 12 (6.4â 26) & --\\
ECM & Short & Illumina & PHYLO & Bray-Curtis & 13 (6.7â 26) & --\\
ECM & Short & PacBio & Cons & Bray-Curtis & 42 ( 16â 60) & --\\
ECM & Short & PacBio & PHYLO & Bray-Curtis & 42 ( 16â 61) & --\\
ECM & Long & PacBio & PHYLO & Bray-Curtis & 14 (5.2â 30) & 4.2 (  2â 19)\\
ECM & Long & PacBio & PHYLO & W-UniFrac & -- & --\\
\bottomrule
\end{tabular}
\end{table}

\FloatBarrier


\end{document}
