# Generate a makefile to 

source("install_packages.R")

library(magrittr)
library(tidyverse)
library(glue)

if (interactive()) {
  base.dir <- str_extract(getwd(), ".+oueme-fungi-transect")
  seq.dir <- file.path(base.dir, "raw_data")
  data.dir <- file.path(base.dir, "data")
  lab.dir <- file.path(data.dir, "lab_setup")
  dataset.file <- file.path(lab.dir, "datasets.csv")
  target <- file.path(base.dir, "demux.make")
} else {
  if (dirname(target) == ".") target <- file.path(base.dir, basename(target))
}

cat("#############################",
    "# Generated by make_demux.R #",
    "# Do not modify!            #",
    "#############################",
    "",
    "# For each sequencing library, constructs a directory which contains",
    "# a separate file for each multiplexed sample (which may be empty).",
    "# The directory also contains a timestamp file, whose recipe actually",
    "# creates each of these files.",
    "# Each of the files also depends on the timestamp file.",
    "# The recipe will be triggered if:",
    "#   - any of the files is missing",
    "#   - any of the files has been modified since the last timestamp",
    "#   - any of the prerequisite files has been modified since the last build",
    "",
    sep = "\n",
    file = target)

datasets <- read_csv(dataset.file) %>%
  mutate(rootdir = file.path(seq.dir, Dataset, Seq.Run), 
         file = map2(rootdir,
                     glue("({Seq.Run}_\\d+\\.reads_of_insert\\.fastq\\.gz|^rawlib.basecaller.bam)"),
                     list.files,
                     recursive = TRUE),
         rootdir = str_replace(rootdir, fixed(seq.dir), "$(SEQDIR)"),
         demux.dir = file.path(rootdir, "demultiplex"))

datasets %<>%
  unnest(file) %>%
  mutate(rootdir = str_replace(rootdir, fixed(seq.dir), "$(SEQDIR)"),
         Stem = str_replace(file, fixed(".reads_of_insert.fastq.gz"), ""),
         Stem = str_replace(Stem, "_?rawlib\\.basecaller\\.bam", ""),
         InFile = str_replace(file, fixed(".bam"), ".fastq.gz"),
         Plate = str_match(Stem, glue("{Seq.Run}_(\\d+)$"))[,2],
         Plate = replace_na(Plate, "001"),
         Plate = glue("{Dataset}-{Plate}"),
         blastdb.fwd = glue("$(TAG_ROOT)/{Forward}"),
         blastdb.rev = glue("$(TAG_ROOT)/{Reverse}"))

#Instructions to demultiplex
#This needs to be done once per sequence library.
datasets %>%
  glue_data("{demux.dir}/.{Plate} : RVARS += blastdb.fwd <- '{blastdb.fwd}'; |",
            "  blastdb.rev <- '{blastdb.rev}'; |",
            "  plate <- '{Plate}'; |",
            "  demux.dir <- '{demux.dir}';",
            "{demux.dir}/.{Plate} : demultiplex_all.R |",
            "  {rootdir}/{InFile} |",
            "  {blastdb.fwd} |",
            "  {blastdb.rev} |",
            "  $(TAG_ROOT)/{Forward}.fasta |",
            "  $(TAG_ROOT)/{Reverse}.fasta |",
            "  $(LABDIR)/{PlateKey}",
            "\t[ -e $(@D) ] || mkdir $(@D)",
            "\trm -f $@.prestamp",
            "\ttouch $@.prestamp",
            "\t$(R)",
            "\tmv -f $@.prestamp $@",
            .sep = "\n") %>%
  glue_collapse(sep = "\n\n") %>%
  str_replace_all(fixed("|\n"), "\\\n") %>%
  cat("\n", sep = "", file = target, append = TRUE)



# List of groups which need to be extracted.
datasets %>%
  mutate(PlateKey = file.path(lab.dir, PlateKey),
         PlateKey = map(PlateKey, read_csv)) %>%
  unnest(PlateKey) %>%
  mutate(OutFile = glue("{file.path(demux.dir, Plate)}_{well}.fastq.gz"),
         TrimFile = str_replace(OutFile,
                                fixed("demultiplex"),
                                "trim"),
         TrimFile = str_replace(TrimFile,
                                fixed(".fastq.gz"),
                                ".trim.fastq.gz")) %>%
  glue_data("demultiplex : {OutFile}",
            "{OutFile} : {demux.dir}/.{Plate} ;",
            "trim : {TrimFile}",
            "{TrimFile} : {OutFile}",
            .sep = "\n",
            .trim = FALSE) %>%
  glue_collapse(sep = "\n\n") %>%
  cat("\n", ., file = target, append = TRUE, sep = "")

  
