# Generate a makefile to 

#source("install_packages.R", echo = TRUE)

library(magrittr)
library(tidyverse)
library(glue)

if (interactive()) {
  base.dir <- str_extract(getwd(), ".+oueme-fungi-transect")
  seq.dir <- file.path(base.dir, "raw_data")
  data.dir <- file.path(base.dir, "data")
  lab.dir <- file.path(data.dir, "lab_setup")
  dataset.file <- file.path(lab.dir, "datasets.csv")
  target <- file.path(base.dir, "demux.make")
  splits <- expand.grid(letters, letters) %>%
    glue_data("x{Var2}{Var1}") %>%
    magrittr::extract(1:4)
} else {
  if (dirname(target) == ".") target <- file.path(base.dir, basename(target))
}

cat("#############################",
    "# Generated by make_demux.R #",
    "# Do not modify!            #",
    "#############################",
    "",
    "# For each sequencing library, constructs a directory which contains",
    "# a separate file for each multiplexed sample (which may be empty).",
    "# The directory also contains a timestamp file, whose recipe actually",
    "# creates each of these files.",
    "# Each of the files also depends on the timestamp file.",
    "# The recipe will be triggered if:",
    "#   - any of the files is missing",
    "#   - any of the files has been modified since the last timestamp",
    "#   - any of the prerequisite files has been modified since the last build",
    "",
    sep = "\n",
    file = target)

# find potential source files
datasets <- read_csv(dataset.file) %>%
  mutate(Regions = map(Regions, ~parse(text = .)) %>%
           map(eval)) %>%
  unnest(Regions) %>%
  mutate(Regions = as.character(Regions) %>%
           paste0(".", .) %>%
           str_replace(fixed(".NULL"), ""),
         dada.file = glue("$(DATADIR)/{Dataset}_{Seq.Run}{Regions}.asv.Rdata"))

datasets %>%
  with(paste0("dada : ", paste0(dada.file, collapse = ' \\\n       '))) %>%
  cat("\n\n", sep = "", file = target, append = TRUE)

datasets %<>%
  mutate(rootdir = file.path(seq.dir, Dataset, Seq.Run), 
         file = map2(rootdir,
                     glue("({Seq.Run}_\\d+\\.reads_of_insert\\.fastq\\.gz|^rawlib.basecaller.bam)"),
                     list.files,
                     recursive = TRUE) %>%
           # don't include any that have already been split
           map(discard,
               str_detect,
               pattern = "-(x[a-z][a-z])\\.fasta\\.gz"),
         rootdir = str_replace(rootdir, fixed(seq.dir), "$(SEQDIR)"),
         demux.dir = file.path(rootdir, "demultiplex")) %>%
  unnest(file) %>%
  mutate(shard = list(splits)) %>%
  unnest(shard) %>%
  mutate(rootdir = str_replace(rootdir, fixed(seq.dir), "$(SEQDIR)"),
         Stem = str_replace(file, fixed(".reads_of_insert.fastq.gz"), ""),
         Stem = str_replace(Stem, "_?rawlib\\.basecaller\\.bam", ""),
         InFile = str_replace(file, fixed(".bam"), ".fastq.gz"),
         InFile = str_replace(InFile, fixed(".fastq.gz"), glue("-{shard}.fastq.gz")),
         Plate = str_match(Stem, glue("{Seq.Run}_(\\d+)$"))[,2],
         Plate = replace_na(Plate, "001"),
         Plate = glue("{Dataset}-{Plate}"),
         blastdb.fwd = glue("$(TAG_ROOT)/{Forward}"),
         blastdb.rev = glue("$(TAG_ROOT)/{Reverse}"))

#Instructions to demultiplex
#This needs to be done once per sequence library.
datasets %>%
  glue_data(".INTERMEDIATE: {demux.dir}/.{Plate}-{shard}",
            "{demux.dir}/.{Plate}-{shard} : RVARS += blastdb.fwd <- '{blastdb.fwd}'; |",
            "  blastdb.rev <- '{blastdb.rev}'; |",
            "  plate <- '{Plate}'; |",
            "  demux.dir <- '{demux.dir}'; |",
            "  shard <- '{shard}';",
            "{demux.dir}/.{Plate}-{shard} : demultiplex_all.R |",
            "  {rootdir}/{InFile} |",
            "  {blastdb.fwd} |",
            "  {blastdb.rev} |",
            "  $(TAG_ROOT)/{Forward}.fasta |",
            "  $(TAG_ROOT)/{Reverse}.fasta |",
            "  $(LABDIR)/{PlateKey}",
            "\t[ -e $(@D) ] || mkdir -p $(@D)",
            "\trm -f $@.prestamp",
            "\ttouch $@.prestamp",
            "\t$(R)",
            "\tmv -f $@.prestamp $@",
            .sep = "\n") %>%
  unique %>%
  glue_collapse(sep = "\n\n") %>%
  str_replace_all(fixed("|\n"), "\\\n") %>%
  cat("\n", sep = "", file = target, append = TRUE)

# List of groups which need to be extracted.
datasets %<>%
  mutate(PlateKey = file.path(lab.dir, PlateKey),
         PlateKey = map(PlateKey, read_csv)) %>%
  unnest(PlateKey) %>%
  mutate(OutFile = glue("{file.path(demux.dir, Plate)}_{well}-{shard}.fastq.gz"))

datasets %>% glue_data("{OutFile} : {demux.dir}/.{Plate}-{shard} ;",
                       .sep = "\n",
                       .trim = FALSE) %>%
  unique %>%
  glue_collapse(sep = "\n\n") %>%
  cat("\n", ., file = target, append = TRUE, sep = "")


datasets %>%
  mutate(demux.file = str_replace(OutFile, fixed(glue("-{shard}")), "")) %>%
  group_by(dada.file, demux.file) %>%
  summarize(shard.file = paste0(OutFile, collapse = " ")) %>%
  glue_data("demultiplex : {demux.file}",
            ".INTERMEDIATE : {shard.file}",
            "{demux.file} : {shard.file}",
            "\t$(UNSPLIT)",
            .sep = "\n",
            .trim = FALSE) %>%
  glue_collapse(sep = "\n\n") %>%
  cat("\n", ., file = target, append = TRUE, sep = "")

datasets %>%
  mutate(region.file = str_replace(OutFile, fixed(glue("-{shard}")),
                                   Regions)) %>%
  select(region.file, dada.file) %>%
  unique %>%
  mutate(trim.file = str_replace(region.file,
                                fixed("demultiplex"),
                                "trim"),
         trim.file = str_replace(trim.file,
                                fixed(".fastq.gz"),
                                ".trim.fastq.gz")) %>%
  glue_data("trim : {trim.file}",
            "{trim.file} : {region.file}",
            "{dada.file} : {trim.file}",
            .sep = "\n",
            .trim = FALSE) %>%
  glue_collapse(sep = "\n\n") %>%
  cat("\n", ., file = target, append = TRUE, sep = "")
